<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • variancePartition</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.35.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/variancePartition.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dream.html">dream analysis</a>
    <a class="dropdown-item" href="../articles/rnd_effects.html">Theory and practice of random effects</a>
    <a class="dropdown-item" href="../articles/mvtests.html">Multivariate tests</a>
    <a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    <a class="dropdown-item" href="../articles/errors.html">Error handling</a>
    <a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.35.3" id="variancepartition-1353">variancePartition 1.35.3<a class="anchor" aria-label="anchor" href="#variancepartition-1353"></a></h2>
<ul><li>June 11, 2024</li>
<li>in <code><a href="../reference/mvTest-method.html">mvTest()</a></code> with <code>shrink.cov = TRUE</code> uses <code>lambda = 0.01</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.35.2" id="variancepartition-1352">variancePartition 1.35.2<a class="anchor" aria-label="anchor" href="#variancepartition-1352"></a></h2>
<ul><li>June 6, 2024</li>
<li>in <code><a href="../reference/mvTest-method.html">mvTest()</a></code> return <code>beta</code> and <code>se</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.35.1" id="variancepartition-1351">variancePartition 1.35.1<a class="anchor" aria-label="anchor" href="#variancepartition-1351"></a></h2>
<ul><li>May 31, 2024</li>
<li>version bump for BioC</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.15" id="variancepartition-13315">variancePartition 1.33.15<a class="anchor" aria-label="anchor" href="#variancepartition-13315"></a></h2>
<ul><li>May 16, 2024</li>
<li>in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>, default changed to <code>rescaleWeightsAfter = FALSE</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.14" id="variancepartition-13314">variancePartition 1.33.14<a class="anchor" aria-label="anchor" href="#variancepartition-13314"></a></h2>
<ul><li>April 19, 2024</li>
<li>in <code><a href="../reference/toptable-method.html">topTable()</a></code> resolve issue when specifying multiple coeffs</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.13" id="variancepartition-13313">variancePartition 1.33.13<a class="anchor" aria-label="anchor" href="#variancepartition-13313"></a></h2>
<ul><li>April 6, 2024</li>
<li>in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>, add argument <code>prior.count.for.weights</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.12" id="variancepartition-13312">variancePartition 1.33.12<a class="anchor" aria-label="anchor" href="#variancepartition-13312"></a></h2>
<ul><li>March 22, 2024</li>
<li>Fix use with <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code> taking raw counts</li>
<li><a href="https://github.com/GabrielHoffman/variancePartition/issues/97" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/97</a></li>
<li>in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>, add argument <code>priorWeightsAsCounts=FALSE</code> and <code>prior.count.for.weights</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.11" id="variancepartition-13311">variancePartition 1.33.11<a class="anchor" aria-label="anchor" href="#variancepartition-13311"></a></h2>
<ul><li>Feb 7, 2024</li>
<li>fix bug in <code>dream(...,ddf="Kenward-Roger")</code> that gave false positives and negatives
<ul><li>scaled weights properly to fix this issue, since <code>df</code> in <code><a href="https://rdrr.io/pkg/lmerTest/man/contest.html" class="external-link">lmerTest::contest()</a></code> assumes mean of weights is 1.</li>
<li>old code used unscaled weights, so df values were too large</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.10" id="variancepartition-13310">variancePartition 1.33.10<a class="anchor" aria-label="anchor" href="#variancepartition-13310"></a></h2>
<ul><li>Feb 5, 2024</li>
<li>in <code><a href="../reference/augmentPriorCount.html">augmentPriorCount()</a></code> and <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>, add argument <code>scaledByLib=FALSE</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.9" id="variancepartition-1339">variancePartition 1.33.9<a class="anchor" aria-label="anchor" href="#variancepartition-1339"></a></h2>
<ul><li>Jan 25, 2024</li>
<li>bug fix in <code><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">BIC()</a></code> and <code>.fitExtractVarPartModel()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.8" id="variancepartition-1338">variancePartition 1.33.8<a class="anchor" aria-label="anchor" href="#variancepartition-1338"></a></h2>
<ul><li>Jan 22, 2024</li>
<li>add <code><a href="https://rdrr.io/r/stats/AIC.html" class="external-link">BIC()</a></code> for result of <code><a href="../reference/dream-method.html">dream()</a></code> and <code>lmFit()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.7" id="variancepartition-1337">variancePartition 1.33.7<a class="anchor" aria-label="anchor" href="#variancepartition-1337"></a></h2>
<ul><li>Jan 19, 2024</li>
<li>add <code><a href="https://rdrr.io/r/stats/logLik.html" class="external-link">logLik()</a></code> for result of <code><a href="../reference/dream-method.html">dream()</a></code> and <code>lmFit()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.6" id="variancepartition-1336">variancePartition 1.33.6<a class="anchor" aria-label="anchor" href="#variancepartition-1336"></a></h2>
<ul><li>fix issue with subsetting using logicals</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.5" id="variancepartition-1335">variancePartition 1.33.5<a class="anchor" aria-label="anchor" href="#variancepartition-1335"></a></h2>
<ul><li>Dec 21, 2023</li>
<li>fix issue with ddf always calling <code>"Satterthwaite"</code>
</li>
<li>enforce subsetting of residuals in <code>assign( "[.MArrayLM2",)</code>
<ul><li>this comes with when <code>rdf &lt; 1</code>
</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.4" id="variancepartition-1334">variancePartition 1.33.4<a class="anchor" aria-label="anchor" href="#variancepartition-1334"></a></h2>
<ul><li>Dec 13, 2023</li>
<li>fix bug with BiocParallel in Windows
<ul><li><a href="https://github.com/GabrielHoffman/variancePartition/issues/91" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/91</a></li>
</ul></li>
<li>handle case where a single contrasts fails in <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.3" id="variancepartition-1333">variancePartition 1.33.3<a class="anchor" aria-label="anchor" href="#variancepartition-1333"></a></h2>
<ul><li>Dec 5, 2023</li>
<li>in <code><a href="https://rdrr.io/r/stats/residuals.html" class="external-link">residuals()</a></code> when dividing by <code>sqrt(1-hatvalues)</code> add small offset to make sure the value is positive</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.2" id="variancepartition-1332">variancePartition 1.33.2<a class="anchor" aria-label="anchor" href="#variancepartition-1332"></a></h2>
<ul><li>Nov 13, 2023</li>
<li>add <code><a href="../reference/augmentPriorCount.html">augmentPriorCount()</a></code>
</li>
<li>add <code>prior.count</code> argument to <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code> and feed it to <code><a href="../reference/augmentPriorCount.html">augmentPriorCount()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.33.1" id="variancepartition-1331">variancePartition 1.33.1<a class="anchor" aria-label="anchor" href="#variancepartition-1331"></a></h2>
<ul><li>Nov 2, 2023</li>
<li>move <code>omp_set_num_threads()</code> deeper in nesting</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.22" id="variancepartition-13122">variancePartition 1.31.22<a class="anchor" aria-label="anchor" href="#variancepartition-13122"></a></h2>
<ul><li>Oct 19, 2023</li>
<li>fix handling of variables with missing data</li>
<li>return <code>fit$genes</code> properly</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.21" id="variancepartition-13121">variancePartition 1.31.21<a class="anchor" aria-label="anchor" href="#variancepartition-13121"></a></h2>
<ul><li>Oct 16, 2023</li>
<li>handle weights properly when the linear mixed model fails for some genes</li>
<li>
<code>lmFit()</code> and</li>
<li>in <code>iterRows()</code> set <code>scale = FALSE</code> as default</li>
<li>in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>, scale in input weights and weights in side <code><a href="../reference/fitVarPartModel-method.html">fitVarPartModel()</a></code>
<ul><li>get weights estimated most similar to <code>voomLmFit()</code>
</li>
</ul></li>
<li>in <code><a href="../reference/dream-method.html">dream()</a></code> use <code>rescaleWeights = FALSE</code> to get <code>sigma</code> estimates compatable with <code>lmFit()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.20" id="variancepartition-13120">variancePartition 1.31.20<a class="anchor" aria-label="anchor" href="#variancepartition-13120"></a></h2>
<ul><li>Sept 26, 2023</li>
<li>allow <code>weights</code> to be a matrix in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.19" id="variancepartition-13119">variancePartition 1.31.19<a class="anchor" aria-label="anchor" href="#variancepartition-13119"></a></h2>
<ul><li>Sept 22, 2023</li>
<li>add <code>rescaleWeightsAfter</code> argument to <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.18" id="variancepartition-13118">variancePartition 1.31.18<a class="anchor" aria-label="anchor" href="#variancepartition-13118"></a></h2>
<ul><li>Sept 5, 2023</li>
<li>improved error handling for <code><a href="../reference/fitVarPartModel-method.html">fitVarPartModel()</a></code>, <code><a href="../reference/fitExtractVarPartModel-method.html">fitExtractVarPartModel()</a></code>, and <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.16" id="variancepartition-13116">variancePartition 1.31.16<a class="anchor" aria-label="anchor" href="#variancepartition-13116"></a></h2>
<ul><li>August 18, 2023</li>
<li>in <code><a href="../reference/dream-method.html">dream()</a></code>, if <code>"Kenward-Roger"</code> is specified but gives covariance matrix that has poor condition number or is not positive definite, then fall back to <code>"Satterthwaite"</code> for hypothesis testing in linear mixed models</li>
<li>Update documentation, and reformat code</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.15" id="variancepartition-13115">variancePartition 1.31.15<a class="anchor" aria-label="anchor" href="#variancepartition-13115"></a></h2>
<ul><li>August 10, 2023</li>
<li>
<code>fit = dream()</code> now returns <code>fit$loglik</code> (the log-likelihood for each gene), and <code>fit$edf</code> (the effective degreees of freedom for each gene)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.13" id="variancepartition-13113">variancePartition 1.31.13<a class="anchor" aria-label="anchor" href="#variancepartition-13113"></a></h2>
<ul><li>August 7, 2023</li>
<li>fix bug in <code><a href="../reference/calcVarPart-method.html">calcVarPart()</a></code> where weights was ignored in some cases</li>
<li>add additional tests to check this</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.12" id="variancepartition-13112">variancePartition 1.31.12<a class="anchor" aria-label="anchor" href="#variancepartition-13112"></a></h2>
<ul><li>July 3, 2023</li>
<li>
<code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code> converts <code>NA</code> <code>contrasts</code> to <code>NULL</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.11" id="variancepartition-13111">variancePartition 1.31.11<a class="anchor" aria-label="anchor" href="#variancepartition-13111"></a></h2>
<ul><li>setting <code>voomWithDreamWeights(..., span="auto")</code> now estimates tuning parameter from data using <code><a href="https://rdrr.io/pkg/fANCOVA/man/loess.as.html" class="external-link">fANCOVA::loess.as()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.10" id="variancepartition-13110">variancePartition 1.31.10<a class="anchor" aria-label="anchor" href="#variancepartition-13110"></a></h2>
<ul><li>
<code>filterInputData()</code> now ensures EList contains a matrix</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.9" id="variancepartition-1319">variancePartition 1.31.9<a class="anchor" aria-label="anchor" href="#variancepartition-1319"></a></h2>
<ul><li>Fix issue in <code><a href="../reference/mvTest-method.html">mvTest()</a></code> when specifying features with strings</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.8" id="variancepartition-1318">variancePartition 1.31.8<a class="anchor" aria-label="anchor" href="#variancepartition-1318"></a></h2>
<ul><li>Fix error message when linear mixed model fails</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.7" id="variancepartition-1317">variancePartition 1.31.7<a class="anchor" aria-label="anchor" href="#variancepartition-1317"></a></h2>
<ul><li>update <code><a href="../reference/mvTest-method.html">mvTest()</a></code> to run in parallel</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.6" id="variancepartition-1316">variancePartition 1.31.6<a class="anchor" aria-label="anchor" href="#variancepartition-1316"></a></h2>
<ul><li>update <code><a href="../reference/mvTest-method.html">mvTest()</a></code> to include Hotelling T2 test and <code>LS.empirical()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.31.1" id="variancepartition-1311">variancePartition 1.31.1<a class="anchor" aria-label="anchor" href="#variancepartition-1311"></a></h2>
<ul><li>June 4, 2023</li>
<li>Rename for Bioconductor compatability</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.5" id="variancepartition-205">variancePartition 2.0.5<a class="anchor" aria-label="anchor" href="#variancepartition-205"></a></h2>
<ul><li>May 31, 2023</li>
<li>fix convergence issues</li>
<li>fix initialization of <code>lmer()</code> fit</li>
<li>use 1 OMP thread internally, then restore to original value</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.4" id="variancepartition-204">variancePartition 2.0.4<a class="anchor" aria-label="anchor" href="#variancepartition-204"></a></h2>
<ul><li>May 30, 2023</li>
<li>When running <code><a href="../reference/dream-method.html">dream()</a></code>, ensure model convergence using second fitting with <code>Nelder_Mead</code> to avoid edge cases where the approximate hessian from <code>lmerTest::as_lmerModLT()</code> has a negative eigenvalue</li>
<li>fix issue in <code><a href="../reference/get_prediction-method.html">get_prediction()</a></code> returning NA values when variables modeled as categorical and levels are omitted</li>
<li>fix issue in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code> when some genes don’t converge</li>
<li>retry <code>lmer()</code> model fit with another optimizer after it fails convergence test.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.3" id="variancepartition-203">variancePartition 2.0.3<a class="anchor" aria-label="anchor" href="#variancepartition-203"></a></h2>
<ul><li>May 13, 2023</li>
<li>fix <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.2" id="variancepartition-202">variancePartition 2.0.2<a class="anchor" aria-label="anchor" href="#variancepartition-202"></a></h2>
<ul><li>May 17, 2023</li>
<li>add matrix argument to <code><a href="../reference/mvTest-method.html">mvTest()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.1" id="variancepartition-201">variancePartition 2.0.1<a class="anchor" aria-label="anchor" href="#variancepartition-201"></a></h2>
<ul><li>May 12, 2023</li>
<li>
<code><a href="../reference/mvTest-method.html">mvTest()</a></code> now shrinks covariance using the Schafer-Strimmer method</li>
<li>
<code><a href="../reference/vcovSqrt-method.html">vcovSqrt()</a></code> returns the matrix whose cross product gives the <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code> result from fits with <code><a href="../reference/dream-method.html">dream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.0" id="variancepartition-200">variancePartition 2.0.0<a class="anchor" aria-label="anchor" href="#variancepartition-200"></a></h2>
<ul><li>April 20, 2023</li>
<li>Major code refactoring to:
<ul><li>improve code reuse</li>
<li>simplify debugging and maintaining code</li>
<li>simplify addition of new features</li>
<li>improve error handling</li>
<li>some linear mixed model analyses are 50% faster</li>
<li>enable additional features for <code>dreamlet</code> package that depends heavily on <code>variancePartition</code>.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.29.2" id="variancepartition-1292">variancePartition 1.29.2<a class="anchor" aria-label="anchor" href="#variancepartition-1292"></a></h2>
<ul><li>Rescale weights</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.9" id="variancepartition-1289">variancePartition 1.28.9<a class="anchor" aria-label="anchor" href="#variancepartition-1289"></a></h2>
<ul><li>March 14, 2023</li>
<li>fix rounding error in <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code>
</li>
<li>add Pearson residuals to <code><a href="https://rdrr.io/r/stats/residuals.html" class="external-link">residuals()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.8" id="variancepartition-1288">variancePartition 1.28.8<a class="anchor" aria-label="anchor" href="#variancepartition-1288"></a></h2>
<ul><li>March 8, 2023</li>
<li>add <code><a href="../reference/mvTest-method.html">mvTest()</a></code> with features as list</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.7" id="variancepartition-1287">variancePartition 1.28.7<a class="anchor" aria-label="anchor" href="#variancepartition-1287"></a></h2>
<ul><li>March 7, 2023</li>
<li>Fix bug in <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code> by adding <code><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.6" id="variancepartition-1286">variancePartition 1.28.6<a class="anchor" aria-label="anchor" href="#variancepartition-1286"></a></h2>
<ul><li>March 1, 2023</li>
<li>
<code><a href="../reference/diffVar-method.html">diffVar()</a></code> now fits contrasts estimated in first step</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.5" id="variancepartition-1285">variancePartition 1.28.5<a class="anchor" aria-label="anchor" href="#variancepartition-1285"></a></h2>
<ul><li>Feb 24, 2023</li>
<li>Fix error in <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code> when samples are dropped due to covariate having <code>NA</code> value</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.4" id="variancepartition-1284">variancePartition 1.28.4<a class="anchor" aria-label="anchor" href="#variancepartition-1284"></a></h2>
<ul><li>Feb 2, 2023</li>
<li>Improve documentation for contrasts</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.3" id="variancepartition-1283">variancePartition 1.28.3<a class="anchor" aria-label="anchor" href="#variancepartition-1283"></a></h2>
<ul><li>Jan 19, 2023</li>
<li>Improve checking and documentation for contrasts</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.2" id="variancepartition-1282">variancePartition 1.28.2<a class="anchor" aria-label="anchor" href="#variancepartition-1282"></a></h2>
<ul><li>Jan 13, 2023</li>
<li>
<code><a href="../reference/canCorPairs.html">canCorPairs()</a></code> now allows random effects in formula
<ul><li>but won’t change results</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.28.1" id="variancepartition-1281">variancePartition 1.28.1<a class="anchor" aria-label="anchor" href="#variancepartition-1281"></a></h2>
<ul><li>Jan 04, 2023</li>
<li>bug fixes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.17" id="variancepartition-12717">variancePartition 1.27.17<a class="anchor" aria-label="anchor" href="#variancepartition-12717"></a></h2>
<ul><li>in <code><a href="../reference/mvTest-method.html">mvTest()</a></code>, more consistent return values when one features is used</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.16" id="variancepartition-12716">variancePartition 1.27.16<a class="anchor" aria-label="anchor" href="#variancepartition-12716"></a></h2>
<ul><li>bug fix</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.15" id="variancepartition-12715">variancePartition 1.27.15<a class="anchor" aria-label="anchor" href="#variancepartition-12715"></a></h2>
<ul><li>fix approximation for <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code>
</li>
<li>add checks for <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code> in <code>test_vcov2()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.14" id="variancepartition-12714">variancePartition 1.27.14<a class="anchor" aria-label="anchor" href="#variancepartition-12714"></a></h2>
<ul><li>fix bug in <code><a href="../reference/toptable-method.html">topTable()</a></code>
</li>
<li>add <code><a href="https://rdrr.io/r/stats/deviance.html" class="external-link">deviance()</a></code>
</li>
<li>update docs</li>
<li>update <code>sqrtMatrix()</code> to have positive diagonal</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.13" id="variancepartition-12713">variancePartition 1.27.13<a class="anchor" aria-label="anchor" href="#variancepartition-12713"></a></h2>
<ul><li>add <code><a href="../reference/diffVar-method.html">diffVar()</a></code> test of differential variance</li>
<li>
<code><a href="../reference/dream-method.html">dream()</a></code> now returns <code>formula</code>, <code>data</code>, and <code>hatvalues</code>
</li>
<li>define <code><a href="https://rdrr.io/r/stats/influence.measures.html" class="external-link">hatvalues()</a></code> for result of <code><a href="../reference/dream-method.html">dream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.12" id="variancepartition-12712">variancePartition 1.27.12<a class="anchor" aria-label="anchor" href="#variancepartition-12712"></a></h2>
<ul><li>in <code><a href="../reference/mvTest-method.html">mvTest()</a></code>:
<ul><li>default method is now <code>"FE"</code>
</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.11" id="variancepartition-12711">variancePartition 1.27.11<a class="anchor" aria-label="anchor" href="#variancepartition-12711"></a></h2>
<ul><li>in <code><a href="../reference/mvTest-method.html">mvTest()</a></code>:
<ul><li>return number of features</li>
<li>return stat.FE and stat.het for RE2C</li>
<li>return NA for stat if 1 feature</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.10" id="variancepartition-12710">variancePartition 1.27.10<a class="anchor" aria-label="anchor" href="#variancepartition-12710"></a></h2>
<ul><li>Add check for very large weights in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>
<ul><li>follows bug report: <a href="https://github.com/GabrielHoffman/variancePartition/issues/66" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/66</a>
</li>
</ul></li>
<li>in <code><a href="../reference/mvTest-method.html">mvTest()</a></code> change option “LS” to “FE”</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.9" id="variancepartition-1279">variancePartition 1.27.9<a class="anchor" aria-label="anchor" href="#variancepartition-1279"></a></h2>
<ul><li>bug fix in <code><a href="../reference/mvTest-method.html">mvTest()</a></code> for contrasts
<ul><li>extensive changes to <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code> in order to handle contrasts</li>
</ul></li>
<li>fix by in subsetting <code>MArrayLM2</code> object. <code>cov.coefficients.list</code> are now named based on gene names</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.8" id="variancepartition-1278">variancePartition 1.27.8<a class="anchor" aria-label="anchor" href="#variancepartition-1278"></a></h2>
<ul><li>bug fix in <code><a href="../reference/mvTest-method.html">mvTest()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.7" id="variancepartition-1277">variancePartition 1.27.7<a class="anchor" aria-label="anchor" href="#variancepartition-1277"></a></h2>
<ul><li>add <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code> for results from <code><a href="../reference/dream-method.html">dream()</a></code>
</li>
<li>add <code><a href="../reference/mvTest-method.html">mvTest()</a></code> for multivariate tests on results from <code><a href="../reference/dream-method.html">dream()</a></code> using <code><a href="https://rdrr.io/r/stats/vcov.html" class="external-link">vcov()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.6" id="variancepartition-1276">variancePartition 1.27.6<a class="anchor" aria-label="anchor" href="#variancepartition-1276"></a></h2>
<ul><li>remove FMT functions</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.5" id="variancepartition-1275">variancePartition 1.27.5<a class="anchor" aria-label="anchor" href="#variancepartition-1275"></a></h2>
<ul><li>update dependencies</li>
<li>make <code><a href="../reference/toptable-method.html">topTable()</a></code> generic to work with R 4.2.1 and Bioc 3.16</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.4" id="variancepartition-1274">variancePartition 1.27.4<a class="anchor" aria-label="anchor" href="#variancepartition-1274"></a></h2>
<ul><li>update dependencies</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.3" id="variancepartition-1273">variancePartition 1.27.3<a class="anchor" aria-label="anchor" href="#variancepartition-1273"></a></h2>
<ul><li>update filtering of covariates, especially for when many samples are dropped</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.27.1" id="variancepartition-1271">variancePartition 1.27.1<a class="anchor" aria-label="anchor" href="#variancepartition-1271"></a></h2>
<ul><li>update <code>plotPercentBars</code> arguments</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.26.1" id="variancepartition-1261">variancePartition 1.26.1<a class="anchor" aria-label="anchor" href="#variancepartition-1261"></a></h2>
<ul><li>patch to update docs</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.13" id="variancepartition-12513">variancePartition 1.25.13<a class="anchor" aria-label="anchor" href="#variancepartition-12513"></a></h2>
<ul><li>Fix compatibility issue with lme4 1.1.29</li>
<li>reported <a href="https://github.com/GabrielHoffman/variancePartition/issues/51" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/51</a>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.12" id="variancepartition-12512">variancePartition 1.25.12<a class="anchor" aria-label="anchor" href="#variancepartition-12512"></a></h2>
<ul><li>in <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code>, fix issue where terms with colon cause and error</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.11" id="variancepartition-12511">variancePartition 1.25.11<a class="anchor" aria-label="anchor" href="#variancepartition-12511"></a></h2>
<ul><li>fix bug in <code><a href="../reference/dream-method.html">dream()</a></code> for variables with NA values</li>
<li>improve handling of invalid contrasts in <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.9" id="variancepartition-1259">variancePartition 1.25.9<a class="anchor" aria-label="anchor" href="#variancepartition-1259"></a></h2>
<ul><li>for <code><a href="../reference/getContrast-method.html">getContrast()</a></code> and <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code> make sure formula argument is a formula and not a string</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.8" id="variancepartition-1258">variancePartition 1.25.8<a class="anchor" aria-label="anchor" href="#variancepartition-1258"></a></h2>
<ul><li>small bug fixes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.7" id="variancepartition-1257">variancePartition 1.25.7<a class="anchor" aria-label="anchor" href="#variancepartition-1257"></a></h2>
<ul><li>
<code><a href="../reference/dream-method.html">dream()</a></code> now drops samples with missing data gracefully</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.6" id="variancepartition-1256">variancePartition 1.25.6<a class="anchor" aria-label="anchor" href="#variancepartition-1256"></a></h2>
<ul><li>fix small plotting bug in <code><a href="../reference/plotStratify.html">plotStratify()</a></code> and <code><a href="../reference/plotStratifyBy.html">plotStratifyBy()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.5" id="variancepartition-1255">variancePartition 1.25.5<a class="anchor" aria-label="anchor" href="#variancepartition-1255"></a></h2>
<ul><li>add <code><a href="../reference/getTreat-method.html">getTreat()</a></code> to evaluate <code>treat()</code>/<code>topTreat()</code> seamlessly on results of <code><a href="../reference/dream-method.html">dream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.4" id="variancepartition-1254">variancePartition 1.25.4<a class="anchor" aria-label="anchor" href="#variancepartition-1254"></a></h2>
<ul><li>in <code><a href="../reference/dream-method.html">dream()</a></code> set default <code>ddf = "adaptive"</code>, which uses “KR” for less than 12 samples</li>
<li>all functions default tp <code>BPPARAM=SerialParam()</code>
</li>
<li>add <code>eBayes()</code> to vignette for <code><a href="../reference/dream-method.html">dream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.3" id="variancepartition-1253">variancePartition 1.25.3<a class="anchor" aria-label="anchor" href="#variancepartition-1253"></a></h2>
<ul><li>add genes argument to <code><a href="../reference/plotPercentBars-method.html">plotPercentBars()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.2" id="variancepartition-1252">variancePartition 1.25.2<a class="anchor" aria-label="anchor" href="#variancepartition-1252"></a></h2>
<ul><li>change <code><a href="../reference/plotPercentBars-method.html">plotPercentBars()</a></code> to use generic S4</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.25.1" id="variancepartition-1251">variancePartition 1.25.1<a class="anchor" aria-label="anchor" href="#variancepartition-1251"></a></h2>
<ul><li>update handling of <code>weights</code> in <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code> and add <code><a href="../reference/applyQualityWeights.html">applyQualityWeights()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.23.6" id="variancepartition-1236">variancePartition 1.23.6<a class="anchor" aria-label="anchor" href="#variancepartition-1236"></a></h2>
<ul><li>update <code><a href="../reference/calcVarPart-method.html">calcVarPart()</a></code> with argument <code>scale=TRUE</code> allowing the user to disable scaling to fractions</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.23.5" id="variancepartition-1235">variancePartition 1.23.5<a class="anchor" aria-label="anchor" href="#variancepartition-1235"></a></h2>
<ul><li>use <code>RhpcBLASctl::omp_set_num_threads(1)</code> to use only 1 OpenMP thread for linear algebra within each BiocParallel process</li>
<li>update <code><a href="../reference/dream-method.html">dream()</a></code> so <code>useWeights=FALSE</code> works with <code>lmFit()</code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.23.4" id="variancepartition-1234">variancePartition 1.23.4<a class="anchor" aria-label="anchor" href="#variancepartition-1234"></a></h2>
<ul><li>convert some warnings to errors</li>
<li>add proper handling of weights to <code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.23.3" id="variancepartition-1233">variancePartition 1.23.3<a class="anchor" aria-label="anchor" href="#variancepartition-1233"></a></h2>
<ul><li>add calcVarPart() support for a range of GLM/GLMM’s</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.23.2" id="variancepartition-1232">variancePartition 1.23.2<a class="anchor" aria-label="anchor" href="#variancepartition-1232"></a></h2>
<ul><li>add flag to checkModelStatus() so warnings are thrown immediately</li>
<li>fix export of as.data.frame</li>
<li>fixed issues where messages were printed even if quiet=TRUE</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.23.1" id="variancepartition-1231">variancePartition 1.23.1<a class="anchor" aria-label="anchor" href="#variancepartition-1231"></a></h2>
<ul><li>New freeze for Bioconductor devel branch</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.10" id="variancepartition-12110">variancePartition 1.21.10<a class="anchor" aria-label="anchor" href="#variancepartition-12110"></a></h2>
<ul><li>add suppressWarnings flag to makeContrastsDream()</li>
<li>ensure that z.std is finite</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.9" id="variancepartition-1219">variancePartition 1.21.9<a class="anchor" aria-label="anchor" href="#variancepartition-1219"></a></h2>
<ul><li>fix bug with makeContrastsDream()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.8" id="variancepartition-1218">variancePartition 1.21.8<a class="anchor" aria-label="anchor" href="#variancepartition-1218"></a></h2>
<ul><li>update dream vignette</li>
<li>update documentation for makeContrastsDream()</li>
<li>fix error in rdf_from_matrices() with eigen() failing</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.7" id="variancepartition-1217">variancePartition 1.21.7<a class="anchor" aria-label="anchor" href="#variancepartition-1217"></a></h2>
<ul><li>Merge changes to contrast code: <a href="https://github.com/GabrielHoffman/variancePartition/pull/32" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/pull/32</a>
</li>
<li>Merge improvments to error checking: <a href="https://github.com/GabrielHoffman/variancePartition/pull/28" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/pull/28</a>
</li>
<li>New warning/error if variables in formula have missing data</li>
<li>add <code><a href="../reference/makeContrastsDream.html">makeContrastsDream()</a></code>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.6" id="variancepartition-1216">variancePartition 1.21.6<a class="anchor" aria-label="anchor" href="#variancepartition-1216"></a></h2>
<ul><li>in dream() allow L to be NULL</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.5" id="variancepartition-1215">variancePartition 1.21.5<a class="anchor" aria-label="anchor" href="#variancepartition-1215"></a></h2>
<ul><li>Fix roxygen2 documentation to pass R CMD check</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.4" id="variancepartition-1214">variancePartition 1.21.4<a class="anchor" aria-label="anchor" href="#variancepartition-1214"></a></h2>
<ul><li>variance fractions for fixed effects model is now computed using new method</li>
<li>fixes subtle issue with previous version where estimates dependend on of terms in the formula</li>
<li><a href="https://github.com/GabrielHoffman/variancePartition/issues/30" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/30</a></li>
<li>Update documentation</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.3" id="variancepartition-1213">variancePartition 1.21.3<a class="anchor" aria-label="anchor" href="#variancepartition-1213"></a></h2>
<ul><li>Faster aggregration after running multiple threads</li>
<li>Pulled from <a href="https://github.com/GabrielHoffman/variancePartition/pull/27" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/pull/27</a>
</li>
<li>by Ryan C. Thompson</li>
<li>eBayes() now works with dream for linear mixed models</li>
<li>add rdf.merMod</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.2" id="variancepartition-1212">variancePartition 1.21.2<a class="anchor" aria-label="anchor" href="#variancepartition-1212"></a></h2>
<ul><li>Reduce size of data passed to each thread by only including variables used in the formula. Applies to multiple functions</li>
<li>add more unit tests</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.21.1" id="variancepartition-1211">variancePartition 1.21.1<a class="anchor" aria-label="anchor" href="#variancepartition-1211"></a></h2>
<ul><li>fix issue with dream and random slopes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.20" id="variancepartition-11920">variancePartition 1.19.20<a class="anchor" aria-label="anchor" href="#variancepartition-11920"></a></h2>
<ul><li>fix bug discovered when the number of features is less than the number of chunks in iterBatch()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.19" id="variancepartition-11919">variancePartition 1.19.19<a class="anchor" aria-label="anchor" href="#variancepartition-11919"></a></h2>
<ul><li>simple bug fixes to pass R CMD check</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.18" id="variancepartition-11918">variancePartition 1.19.18<a class="anchor" aria-label="anchor" href="#variancepartition-11918"></a></h2>
<ul><li>simplify calcVarPart for lm and lmer. Add compatibility for glm</li>
<li>Simplify checkModelStatus.merMod to allow formula (A|B) where A is continuous</li>
<li>remove unused “adjust” arguments for clarity</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.17" id="variancepartition-11917">variancePartition 1.19.17<a class="anchor" aria-label="anchor" href="#variancepartition-11917"></a></h2>
<ul><li>add get_prediction() for results of lm()</li>
<li>improve documentation of get_prediction()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.16" id="variancepartition-11916">variancePartition 1.19.16<a class="anchor" aria-label="anchor" href="#variancepartition-11916"></a></h2>
<ul><li>in canCorPairs() change statistic used to summarize CCA to Cramer’s V. The difference is very subtle, but is now based on first principles.</li>
<li>in dream, check that data is a data.frame</li>
<li>dream() defaults to computeResiduals=TRUE for compatability with zenith</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.14" id="variancepartition-11914">variancePartition 1.19.14<a class="anchor" aria-label="anchor" href="#variancepartition-11914"></a></h2>
<ul><li>fix issue with residuals() where examples fail</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.13" id="variancepartition-11913">variancePartition 1.19.13<a class="anchor" aria-label="anchor" href="#variancepartition-11913"></a></h2>
<ul><li>fix issues with residuals()</li>
<li><a href="https://github.com/GabrielHoffman/variancePartition/issues/18" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/18</a></li>
<li>fix issue exporting eBayes, topTable, etc</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.12" id="variancepartition-11912">variancePartition 1.19.12<a class="anchor" aria-label="anchor" href="#variancepartition-11912"></a></h2>
<ul><li>Improve documentation for contrasts in dream.Rmd</li>
<li>check that contrasts sum to zero in plotContrasts.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.11" id="variancepartition-11911">variancePartition 1.19.11<a class="anchor" aria-label="anchor" href="#variancepartition-11911"></a></h2>
<ul><li>in voomWithDreamWeights() fix issue with not defining design</li>
<li><a href="https://github.com/GabrielHoffman/variancePartition/issues/17" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/17</a></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.10" id="variancepartition-11910">variancePartition 1.19.10<a class="anchor" aria-label="anchor" href="#variancepartition-11910"></a></h2>
<ul><li>in voomWithDreamWeights() fix issue with returning design matrix</li>
<li>better error if counts can’t be converted to matrix</li>
<li><a href="https://github.com/GabrielHoffman/variancePartition/issues/15" class="external-link uri">https://github.com/GabrielHoffman/variancePartition/issues/15</a></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.7" id="variancepartition-1197">variancePartition 1.19.7<a class="anchor" aria-label="anchor" href="#variancepartition-1197"></a></h2>
<ul><li>Round numbers in plotContrasts()</li>
<li>fix issues with strings are passed to formula arguments</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.6" id="variancepartition-1196">variancePartition 1.19.6<a class="anchor" aria-label="anchor" href="#variancepartition-1196"></a></h2>
<ul><li>New gives meaning full error message for dream(), etc when variable is not found in data.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.5" id="variancepartition-1195">variancePartition 1.19.5<a class="anchor" aria-label="anchor" href="#variancepartition-1195"></a></h2>
<ul><li>Better error catching when running fitVarPartModel() with fxn that fails</li>
<li>add get_prediction() function</li>
<li>the following code now can be run in parallel fitList = fitVarPartModel( Y, ~ (1|Batch), data, fxn = function(fit){ B = variancePartition::get_prediction(fit, ~(1|Batch)) <a href="mailto:fit@resp" class="email">fit@resp</a>$y - B }, BPPARAM=SnowParam(3))</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.4" id="variancepartition-1194">variancePartition 1.19.4<a class="anchor" aria-label="anchor" href="#variancepartition-1194"></a></h2>
<ul><li>Update vignette <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/issues/3" class="external-link">#3</a>, and update documentation of REML argument</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.3" id="variancepartition-1193">variancePartition 1.19.3<a class="anchor" aria-label="anchor" href="#variancepartition-1193"></a></h2>
<ul><li>add new FAQ.Rmd</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.2" id="variancepartition-1192">variancePartition 1.19.2<a class="anchor" aria-label="anchor" href="#variancepartition-1192"></a></h2>
<ul><li>canCorPairs() now returns NA correlation when two variables have no overlapping observed values</li>
<li>plotCorrMatrix() now handles NA correlation values</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.19.1" id="variancepartition-1191">variancePartition 1.19.1<a class="anchor" aria-label="anchor" href="#variancepartition-1191"></a></h2>
<ul><li>Bump to next Bioconductor version</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.18.3" id="variancepartition-1183">variancePartition 1.18.3<a class="anchor" aria-label="anchor" href="#variancepartition-1183"></a></h2>
<ul><li>Improve documentation</li>
<li>move location of eBayesFMT code</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.18.2" id="variancepartition-1182">variancePartition 1.18.2<a class="anchor" aria-label="anchor" href="#variancepartition-1182"></a></h2>
<ul><li>Clean up some code and add documentation</li>
<li>document ebayesFMT</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.18.1" id="variancepartition-1181">variancePartition 1.18.1<a class="anchor" aria-label="anchor" href="#variancepartition-1181"></a></h2>
<ul><li>Clean up some code and add documentation</li>
<li>compute effective degrees of freedom for each model</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.18.0" id="variancepartition-1180">variancePartition 1.18.0<a class="anchor" aria-label="anchor" href="#variancepartition-1180"></a></h2>
<ul><li>Bioconductor freeze</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.10" id="variancepartition-11710">variancePartition 1.17.10<a class="anchor" aria-label="anchor" href="#variancepartition-11710"></a></h2>
<ul><li>fix issue returning residuals from limma</li>
<li>resolve issue where dream gives error: r[cbind(1L:p, 1L:p)] &lt;- 1 : subscript out of bounds</li>
<li>only occured when no fixed effects were used</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.9" id="variancepartition-1179">variancePartition 1.17.9<a class="anchor" aria-label="anchor" href="#variancepartition-1179"></a></h2>
<ul><li>Fix issues with compatability with R/4.0.0</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.8" id="variancepartition-1178">variancePartition 1.17.8<a class="anchor" aria-label="anchor" href="#variancepartition-1178"></a></h2>
<ul><li>Better error message when response contains missing data</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.7" id="variancepartition-1177">variancePartition 1.17.7<a class="anchor" aria-label="anchor" href="#variancepartition-1177"></a></h2>
<ul><li>Better error message when variable in formula does not exist</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.6" id="variancepartition-1176">variancePartition 1.17.6<a class="anchor" aria-label="anchor" href="#variancepartition-1176"></a></h2>
<ul><li>comply with new Bioconductor check: <em>R_CHECK_LENGTH_1_LOGIC2</em>
</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.5" id="variancepartition-1175">variancePartition 1.17.5<a class="anchor" aria-label="anchor" href="#variancepartition-1175"></a></h2>
<ul><li>if contrasts for dream() is data.frame, convert to matrix</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.4" id="variancepartition-1174">variancePartition 1.17.4<a class="anchor" aria-label="anchor" href="#variancepartition-1174"></a></h2>
<ul><li>Don’t print warnings for residuals() when only one argument passed.</li>
<li>fix bug with residuals evaluated with only fixed effects</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.3" id="variancepartition-1173">variancePartition 1.17.3<a class="anchor" aria-label="anchor" href="#variancepartition-1173"></a></h2>
<ul><li>Allow sparseMatrix for gene expression. Now saves memory by avoiding conversion to matrix. Processing sparseMatrix will be slower, but memory usage will be low.</li>
<li>dream(…, computeResiduals=TRUE) now computes residuals and allows use of residuals() function</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.2" id="variancepartition-1172">variancePartition 1.17.2<a class="anchor" aria-label="anchor" href="#variancepartition-1172"></a></h2>
<ul><li>fix error in voomWithDreamWeights() when design matrix is null</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.17.1" id="variancepartition-1171">variancePartition 1.17.1<a class="anchor" aria-label="anchor" href="#variancepartition-1171"></a></h2>
<ul><li>topTable(…,sort.by=) now is correct when and F-test is used</li>
<li>fixed issue in classifyTestsF.MArrayLM2, now is much faster</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.8" id="variancepartition-1158">variancePartition 1.15.8<a class="anchor" aria-label="anchor" href="#variancepartition-1158"></a></h2>
<ul><li>Replace cat() with message()</li>
<li>add quiet option to a few functions</li>
<li>dream() does not call eBayes() when lmFit is used</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.7" id="variancepartition-1157">variancePartition 1.15.7<a class="anchor" aria-label="anchor" href="#variancepartition-1157"></a></h2>
<ul><li>Official release to development branch</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.6" id="variancepartition-1156">variancePartition 1.15.6<a class="anchor" aria-label="anchor" href="#variancepartition-1156"></a></h2>
<ul><li>fix convergence errror when recycling parameters values from first gene</li>
<li>add column z.std and F.std to topTable</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.4" id="variancepartition-1154">variancePartition 1.15.4<a class="anchor" aria-label="anchor" href="#variancepartition-1154"></a></h2>
<ul><li>Add error message when scale of fixed effects causes a problem</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.3" id="variancepartition-1153">variancePartition 1.15.3<a class="anchor" aria-label="anchor" href="#variancepartition-1153"></a></h2>
<ul><li>Try changing order of eBayes</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.2" id="variancepartition-1152">variancePartition 1.15.2<a class="anchor" aria-label="anchor" href="#variancepartition-1152"></a></h2>
<ul><li>Update vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.15.0" id="variancepartition-1150">variancePartition 1.15.0<a class="anchor" aria-label="anchor" href="#variancepartition-1150"></a></h2>
<ul><li>Push changes to Bioconductor devel</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.11" id="variancepartition-11311">variancePartition 1.13.11<a class="anchor" aria-label="anchor" href="#variancepartition-11311"></a></h2>
<ul><li>Fixed bug in voomWithDreamWeights() with approxfun() with ties</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.10" id="variancepartition-11310">variancePartition 1.13.10<a class="anchor" aria-label="anchor" href="#variancepartition-11310"></a></h2>
<ul><li>fix parallel processing issues</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.9" id="variancepartition-1139">variancePartition 1.13.9<a class="anchor" aria-label="anchor" href="#variancepartition-1139"></a></h2>
<ul><li>update vignette for new parallel processing backend</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.8" id="variancepartition-1138">variancePartition 1.13.8<a class="anchor" aria-label="anchor" href="#variancepartition-1138"></a></h2>
<ul><li>apply empirical Bayes when doing F-test</li>
<li>hypothesis testing for single coefficients is now included by default, so only need to specify contrast matrix if for more complicated contrasts<br></li>
<li>add voomWithDreamWeights() for computing observation weights using random effects</li>
<li>Add BiocParallel capability with BPPARAM argument</li>
<li>allows parallel processing with lower memory usage</li>
<li>dream() is now compatable with gene set enrichments from pinnacle (software comming soon)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.7" id="variancepartition-1137">variancePartition 1.13.7<a class="anchor" aria-label="anchor" href="#variancepartition-1137"></a></h2>
<ul><li>in dream(), add support for genes annotation in DGElist()</li>
<li>in dream(), automatically evaluate contrasts for all single coefficients</li>
<li>add future compatability for gene set enrichments method “pinnacle”</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.5" id="variancepartition-1135">variancePartition 1.13.5<a class="anchor" aria-label="anchor" href="#variancepartition-1135"></a></h2>
<ul><li>add plotContrasts() to dream vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.4" id="variancepartition-1134">variancePartition 1.13.4<a class="anchor" aria-label="anchor" href="#variancepartition-1134"></a></h2>
<ul><li>export classes to fix bug with class “varPartResults” not being defined</li>
<li>Thanks Megan Behringer</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.3" id="variancepartition-1133">variancePartition 1.13.3<a class="anchor" aria-label="anchor" href="#variancepartition-1133"></a></h2>
<ul><li>add plotContrasts()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.2" id="variancepartition-1132">variancePartition 1.13.2<a class="anchor" aria-label="anchor" href="#variancepartition-1132"></a></h2>
<ul><li>Enable random slope models in dream, but not for estimating variance fractions</li>
<li>Thanks Jonas Zierer</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.13.1" id="variancepartition-1131">variancePartition 1.13.1<a class="anchor" aria-label="anchor" href="#variancepartition-1131"></a></h2>
<ul><li>
<h2 id="add-progress-bar-at-eta" class="pkg-version" data-toc-text="1.13.1">Add progress bar at ETA</h2>
<h2 id="variancepartition-1124" class="pkg-version" data-toc-text="1.13.1">variancePartition 1.12.4</h2>
</li>
<li>Update dream vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.12.3" id="variancepartition-1123">variancePartition 1.12.3<a class="anchor" aria-label="anchor" href="#variancepartition-1123"></a></h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>o add <span class="fu">plotContrasts</span>()</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>o Enable random slope models <span class="cf">in</span> dream, but not <span class="cf">for</span> estimating variance fractions</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>o export classes to fix bug with class <span class="st">"varPartResults"</span> not being defined</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a> <span class="sc">-</span> Thanks Megan Behringer</span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.13" id="variancepartition-11113">variancePartition 1.11.13<a class="anchor" aria-label="anchor" href="#variancepartition-11113"></a></h2>
<ul><li>Fix multithreading issue</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.11" id="variancepartition-11111">variancePartition 1.11.11<a class="anchor" aria-label="anchor" href="#variancepartition-11111"></a></h2>
<ul><li>dream can handle multiple contrasts at the same time</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.10" id="variancepartition-11110">variancePartition 1.11.10<a class="anchor" aria-label="anchor" href="#variancepartition-11110"></a></h2>
<ul><li>fix typos in dream vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.8" id="variancepartition-1118">variancePartition 1.11.8<a class="anchor" aria-label="anchor" href="#variancepartition-1118"></a></h2>
<ul><li>Check and stop() if response variable has variance of 0</li>
<li>in dream(), fitExtractVarPartModel(), and fitVarPartModel()</li>
<li>add standardized_t_stat() implicitly in eBayes() using MArrayLM2 class</li>
<li>this transforms moderated t-statistics to have same degrees of freedom</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.7" id="variancepartition-1117">variancePartition 1.11.7<a class="anchor" aria-label="anchor" href="#variancepartition-1117"></a></h2>
<ul><li>Simplify object return by dream to be more more similar to lmFit</li>
<li>now returns MArrayLM instead of MArrayLMM_lmer</li>
<li>if a fixed effects formula is specified (i.e. not random terms)</li>
<li>dream call lmFit in the backend</li>
<li>getContrast() works seamlessly</li>
<li>dream() now returns gene annotation if passed to function</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.6" id="variancepartition-1116">variancePartition 1.11.6<a class="anchor" aria-label="anchor" href="#variancepartition-1116"></a></h2>
<ul><li>add error checing for L in dream</li>
<li>fix typoes in dream vignette</li>
<li>fix typoes in theory_practice_random_effects.Rnw</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.5" id="variancepartition-1115">variancePartition 1.11.5<a class="anchor" aria-label="anchor" href="#variancepartition-1115"></a></h2>
<ul><li>Add dream function for differential expression for repeated measures with a linear mixed model</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.2" id="variancepartition-1112">variancePartition 1.11.2<a class="anchor" aria-label="anchor" href="#variancepartition-1112"></a></h2>
<ul><li>Add warnings to canCorPairs for colinear terms</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.11.1" id="variancepartition-1111">variancePartition 1.11.1<a class="anchor" aria-label="anchor" href="#variancepartition-1111"></a></h2>
<ul><li>Add vignette: theory_practice_random_effects.Rnw</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.9.9" id="variancepartition-199">variancePartition 1.9.9<a class="anchor" aria-label="anchor" href="#variancepartition-199"></a></h2>
<ul><li>Fix issue when package is autoloaded when starting R</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.9.8" id="variancepartition-198">variancePartition 1.9.8<a class="anchor" aria-label="anchor" href="#variancepartition-198"></a></h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a> o Fix issue where <span class="cf">if</span> info data.frame contained a column name <span class="st">"gene"</span>, </span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="fu">fitExtractVarPartModel</span>() would not run</span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.9.6" id="variancepartition-196">variancePartition 1.9.6<a class="anchor" aria-label="anchor" href="#variancepartition-196"></a></h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>    o Fix tximport issue with eval<span class="ot">=</span><span class="cn">FALSE</span></span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.9.2" id="variancepartition-192">variancePartition 1.9.2<a class="anchor" aria-label="anchor" href="#variancepartition-192"></a></h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>    o Fix vignette</span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.9.1" id="variancepartition-191">variancePartition 1.9.1<a class="anchor" aria-label="anchor" href="#variancepartition-191"></a></h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>    o Fix formatting of vignette</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>    o add description of <span class="fu">canCorPairs</span>() <span class="cf">function</span></span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.7" id="variancepartition-157">variancePartition 1.5.7<a class="anchor" aria-label="anchor" href="#variancepartition-157"></a></h2>
<ul><li>include splines in foreach .packages</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.6" id="variancepartition-156">variancePartition 1.5.6<a class="anchor" aria-label="anchor" href="#variancepartition-156"></a></h2>
<ul><li>compatibility with tximport v1.3.5</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.5" id="variancepartition-155">variancePartition 1.5.5<a class="anchor" aria-label="anchor" href="#variancepartition-155"></a></h2>
<ul><li>Decrease computing time of effective sample size with ESS() by additional ~10x with sparse solver</li>
<li>fix margins for plotPercentBars()</li>
<li>Fix bug for getVarianceComponents() when correlated continous variables are included</li>
<li>compatibility with ggplot2 2.2.0</li>
<li>center plot titles</li>
<li>fix order of bars in plotPercentBars()</li>
<li>legend background to transparent</li>
<li>set text to be black</li>
<li>include lme4 in foreach .packages</li>
<li>change residuals color to not be transparent</li>
<li>add CITATION information</li>
<li>plotCorrMatrix now shows dendrogram by default</li>
<li>Estimate run time for fitExtractVarPartModel() / fitVarPartModel()</li>
<li>improve warnings for plotPercentBar()</li>
<li>improve warnings for plotCorrStructure()</li>
<li>define ylab for plotVarPart()</li>
<li>add as.matrix.varPartResults() (hidden)</li>
<li>define isVaryingCoefficientModel() (hidden)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.11" id="variancepartition-1311-1">variancePartition 1.3.11<a class="anchor" aria-label="anchor" href="#variancepartition-1311-1"></a></h2>
<ul><li>in canCorPairs() and other functions, convert formula with as.formula()</li>
<li>improve error messages for canCorPairs()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.10" id="variancepartition-1310">variancePartition 1.3.10<a class="anchor" aria-label="anchor" href="#variancepartition-1310"></a></h2>
<ul><li>Add plotStratify()</li>
<li>Update documentation</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.8" id="variancepartition-138">variancePartition 1.3.8<a class="anchor" aria-label="anchor" href="#variancepartition-138"></a></h2>
<ul><li>Add additional examples to vignette</li>
<li>show projected memory usage of fitVarPartModel()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.7" id="variancepartition-137">variancePartition 1.3.7<a class="anchor" aria-label="anchor" href="#variancepartition-137"></a></h2>
<ul><li>fitVarPartModel warns if names in exprObj and data are not identical</li>
<li>residuals() and other functions deal with missing values properly</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.6" id="variancepartition-136">variancePartition 1.3.6<a class="anchor" aria-label="anchor" href="#variancepartition-136"></a></h2>
<ul><li>Small changes to vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.5" id="variancepartition-135">variancePartition 1.3.5<a class="anchor" aria-label="anchor" href="#variancepartition-135"></a></h2>
<ul><li>Fix Bioconductor error</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.4" id="variancepartition-134">variancePartition 1.3.4<a class="anchor" aria-label="anchor" href="#variancepartition-134"></a></h2>
<ul><li>Fix typos</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.3" id="variancepartition-133">variancePartition 1.3.3<a class="anchor" aria-label="anchor" href="#variancepartition-133"></a></h2>
<ul><li>Improve documentation</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.9" id="variancepartition-119">variancePartition 1.1.9<a class="anchor" aria-label="anchor" href="#variancepartition-119"></a></h2>
<ul><li>Update sortCols to handle Measurement.error</li>
<li>change backend package structure</li>
<li>set Residuals to be grey by default in plotVarPart() and plotPercentBars()</li>
<li>add control = lme4::lmerControl(calc.derivs=FALSE, check.rankX=“stop.deficient” )</li>
<li>add plotCorrStructure</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.8" id="variancepartition-118">variancePartition 1.1.8<a class="anchor" aria-label="anchor" href="#variancepartition-118"></a></h2>
<ul><li>Add ESS.R</li>
<li>Add fitVarTest.R</li>
<li>use lmerTest by default</li>
<li>fix bug checkModelStatus() for variables with backticks in name</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.7" id="variancepartition-117">variancePartition 1.1.7<a class="anchor" aria-label="anchor" href="#variancepartition-117"></a></h2>
<ul><li>GPL License</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.6" id="variancepartition-116">variancePartition 1.1.6<a class="anchor" aria-label="anchor" href="#variancepartition-116"></a></h2>
<ul><li>Move packages from Depends to Imports</li>
<li>For clarity, replace = with &lt;- in parts of examples and vignette</li>
<li>Stop cluster in examples to solve error on Windows machines</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.5" id="variancepartition-115">variancePartition 1.1.5<a class="anchor" aria-label="anchor" href="#variancepartition-115"></a></h2>
<ul><li>Stop cluster in vignette to solve error on Windows machines</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.4" id="variancepartition-114">variancePartition 1.1.4<a class="anchor" aria-label="anchor" href="#variancepartition-114"></a></h2>
<ul><li>Fix Bioconductor check error</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.3" id="variancepartition-113">variancePartition 1.1.3<a class="anchor" aria-label="anchor" href="#variancepartition-113"></a></h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>o Add details to vignette</span></code></pre></div>
<ul><li>Fix ggplot2 compatibility issues</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.2" id="variancepartition-112">variancePartition 1.1.2<a class="anchor" aria-label="anchor" href="#variancepartition-112"></a></h2>
<ul><li>Add details to vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.1" id="variancepartition-111">variancePartition 1.1.1<a class="anchor" aria-label="anchor" href="#variancepartition-111"></a></h2>
<ul><li>add plotPercentBars() to vizualize variance fractions for a subset of genes</li>
<li>add ESS() to compute effective sample size</li>
<li>fix x.labels argument in plotStratifyBy(). Previously, this argument was not used correctly</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.0" id="variancepartition-100">variancePartition 1.0.0<a class="anchor" aria-label="anchor" href="#variancepartition-100"></a></h2>
<ul><li>Release to Bioconductor 3.2</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.9" id="variancepartition-0999">variancePartition 0.99.9<a class="anchor" aria-label="anchor" href="#variancepartition-0999"></a></h2>
<ul><li>add legend argument to plotStratifyBy()</li>
<li>improve warnings / errors for varying coefficient models</li>
<li>allow user to manually adjust cutoff for determining when design matrix is singular
<ul><li>changed default cutoff to 0.999 from 0.99</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.8" id="variancepartition-0998">variancePartition 0.99.8<a class="anchor" aria-label="anchor" href="#variancepartition-0998"></a></h2>
<ul><li>improve warnings / errors when design matrix is close to or exactly singular</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.7" id="variancepartition-0997">variancePartition 0.99.7<a class="anchor" aria-label="anchor" href="#variancepartition-0997"></a></h2>
<ul><li>added new class varPartResults to store results of fitExtractVarPartModel() and extractVarPart()
<ul><li>the user will not notice any change, only the backend is different o Allow computation of adjusted ICC in addition to ICC.</li>
</ul></li>
<li>add warning when categorical variables are modeled as fixed effects</li>
<li>fix computation of variance fractions for varying coefficient models</li>
<li>add getVarianceComponents() to return variances from lmer() or lm() model fit</li>
<li>showWarnings=FALSE suppresses warning messages</li>
<li>add fxn argument to fitVarPartModel to evaluate any function on the model fit</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.6" id="variancepartition-0996">variancePartition 0.99.6<a class="anchor" aria-label="anchor" href="#variancepartition-0996"></a></h2>
<ul><li>Update DESCRIPTION information</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.5" id="variancepartition-0995">variancePartition 0.99.5<a class="anchor" aria-label="anchor" href="#variancepartition-0995"></a></h2>
<ul><li>residuals deals with missing data gracefully and returns a matrix</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.4" id="variancepartition-0994">variancePartition 0.99.4<a class="anchor" aria-label="anchor" href="#variancepartition-0994"></a></h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>   o add documentation <span class="cf">for</span> example datasets</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>   o convert <span class="fu">calcVarPart</span>() to S4 from S3 <span class="cf">function</span> call</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>   o fix typos <span class="cf">in</span> vignette</span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.3" id="variancepartition-0993">variancePartition 0.99.3<a class="anchor" aria-label="anchor" href="#variancepartition-0993"></a></h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>   o <span class="fu">fitVarPartModel</span>() and <span class="fu">fitExractVarPartModel</span>() use S4 instead of S3 calls </span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.2" id="variancepartition-0992">variancePartition 0.99.2<a class="anchor" aria-label="anchor" href="#variancepartition-0992"></a></h2>
<ul><li>rename sort.varParFrac to sortCols</li>
<li>support ExpressionSet</li>
<li>change options for plotStratifyBy()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.12" id="variancepartition-0012">variancePartition 0.0.12<a class="anchor" aria-label="anchor" href="#variancepartition-0012"></a></h2>
<ul><li>add plotStratifyBy()</li>
<li>update documentation</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.11" id="variancepartition-0011">variancePartition 0.0.11<a class="anchor" aria-label="anchor" href="#variancepartition-0011"></a></h2>
<ul><li>fix sort to work on correct argument</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.10" id="variancepartition-0010">variancePartition 0.0.10<a class="anchor" aria-label="anchor" href="#variancepartition-0010"></a></h2>
<ul><li>fitExtractVarPartModel() and fitVarPartModel() now take subset argument</li>
<li>throw warning when no Intercept is specified</li>
<li>if using lmer, warning if categorical variable is modeled as fixed effect</li>
<li>fixed calcVarPart bug with reporting too few variances for multicategory fixed effects</li>
<li>add colinearityScore</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.9" id="variancepartition-009">variancePartition 0.0.9<a class="anchor" aria-label="anchor" href="#variancepartition-009"></a></h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>o <span class="cf">function</span> now use the precision weights when specified</span></code></pre></div>
<ul><li>remove warning about unspecified weights, when useWeights=TRUE</li>
<li>fix issue with sort with only one variable</li>
<li>add main argument to plotVarPart</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.8" id="variancepartition-008">variancePartition 0.0.8<a class="anchor" aria-label="anchor" href="#variancepartition-008"></a></h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>    o Bug update to vignette and simulated data</span></code></pre></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.7" id="variancepartition-007">variancePartition 0.0.7<a class="anchor" aria-label="anchor" href="#variancepartition-007"></a></h2>
<ul><li>remove ‘variable’ from xlab of plotVarPart</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.6" id="variancepartition-006">variancePartition 0.0.6<a class="anchor" aria-label="anchor" href="#variancepartition-006"></a></h2>
<ul><li>set REML=FALSE to default. This fixes issues of inaccurate variance estiamtes, and makes lmer() results more concordant with lm() results</li>
<li>Fix residuals function when lm or lmer is used</li>
<li>fix useWeights argument error for fitExtractVarPartModel()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.5" id="variancepartition-005">variancePartition 0.0.5<a class="anchor" aria-label="anchor" href="#variancepartition-005"></a></h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>    o <span class="fu">plotVarPart</span>() as ylim<span class="ot">=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">100</span>) as default, and can be changed by user</span></code></pre></div>
<ul><li>fitExtractVarPartModel() labels rows correctly</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.4" id="variancepartition-004">variancePartition 0.0.4<a class="anchor" aria-label="anchor" href="#variancepartition-004"></a></h2>
<ul><li>Add sort() for output of extractVarPart() or fitExtractVarPartModel()</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.0.3" id="variancepartition-003">variancePartition 0.0.3<a class="anchor" aria-label="anchor" href="#variancepartition-003"></a></h2>
<ul><li>add residuals function</li>
<li>add fitExtractVarPartModel()</li>
<li>foreach loops us iterators</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.99.0" id="variancepartition-0990">variancePartition 0.99.0<a class="anchor" aria-label="anchor" href="#variancepartition-0990"></a></h2>
<ul><li>Initial version</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

