<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>dream analysis • variancePartition</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="dream analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.36.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/variancePartition.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/dream.html">dream analysis</a></li>
    <li><a class="dropdown-item" href="../articles/rnd_effects.html">Theory and practice of random effects</a></li>
    <li><a class="dropdown-item" href="../articles/mvtests.html">Multivariate tests</a></li>
    <li><a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>dream analysis</h1>
            <h3 data-toc-skip class="subtitle">Differential expression
testing with linear mixed models for repeated measures</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2024-11-05
15:49:29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/vignettes/dream.Rmd" class="external-link"><code>vignettes/dream.Rmd</code></a></small>
      <div class="d-none name"><code>dream.Rmd</code></div>
    </div>

    
    
<!---
cd /Users/gabrielhoffman/workspace/repos
R
library('knitr')
rmarkdown::render('variancePartition/vignettes/dream.Rmd')
rmarkdown::render('dream.Rmd')

# create public version
grep -v "## Warning" /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes/dream.html > /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes/dream2.html
 scp /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes/dream2.html minerva:/hpc/users/hoffmg01/www/software/dream.html
--->
<style>
body {
text-align: justify}
</style>
<!---
# Description
Dream uses a linear model model to increase power and decrease false positives for RNA-seq datasets with repeated measurements.  Dream achieves this by combining multiple statistical concepts into a single statistical model.  The dream model includes:     


| Model property                                                 | Package           | Reference                           |
|:---------------------------------------------------------------|:------------------|:------------------------------------|
| precision weights to model measurement error in RNA-seq counts | limma/voom        | @Law2014                            |
| ability to model multiple random effects                       | lme4              | @Bates2015                          |
| random effects estimated separately for each gene              | variancePartition | @hoffman2016                        |
| hypothesis testing for fixed effects in linear mixed models    | lmerTest          | Kuznetsova, et al. -@kuznetsova2017 |
| small sample size hypothesis test                              | pbkrtest          | Halekoh, et al. -@halekoh2014       |
|                                                                |                   |                                     |
* flexible modeling of repeated measures gene expression data
* precision weights to model measurement error in RNA-seq counts
* ability to model multiple random effects
* random effects estimated separately for each gene
* hypothesis testing for fixed effects in linear mixed models
* small sample size hypothesis test
* multi-threaded analysis across thousands of genes on a multi-core machine.
---><p><strong>D</strong>ifferential expression for
<strong>re</strong>pe<strong>a</strong>ted <strong>m</strong>easures
(dream) uses a linear model model to increase power and decrease false
positives for RNA-seq datasets with multiple measurements per
individual. The analysis fits seamlessly into the widely used workflow
of limma/voom <span class="citation">(Law et al. 2014)</span>. Dream
uses a linear model model to increase power and decrease false positives
for RNA-seq datasets with repeated measurements. Dream achieves this by
combining multiple statistical concepts into a single statistical model.
The model includes:</p>
<ul>
<li>flexible modeling of repeated measures gene expression data<br>
</li>
<li>precision weights to model measurement error in RNA-seq
counts<br>
</li>
<li>ability to model multiple random effects<br>
</li>
<li>random effects estimated separately for each gene<br>
</li>
<li>hypothesis testing for fixed effects in linear mixed models<br>
</li>
<li>small sample size hypothesis test<br>
</li>
</ul>
<p>Dream also includes multi-threaded analysis across thousands of genes
on a multi-core machine.</p>
<div class="section level2">
<h2 id="standard-rna-seq-processing">Standard RNA-seq processing<a class="anchor" aria-label="anchor" href="#standard-rna-seq-processing"></a>
</h2>
<p>This tutorial assumes that the reader is familiar with the limma/voom
workflow for RNA-seq. Process raw count data using limma/voom.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">"variancePartition"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">"edgeR"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">"BiocParallel"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">varPartDEdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter genes by number of counts</span></span>
<span><span class="va">isexpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span></span>
<span></span>
<span><span class="co"># Standard usage of limma/voom</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">[</span><span class="va">isexpr</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make this vignette faster by analyzing a subset of genes</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="va">dge</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="limma-analysis">Limma Analysis<a class="anchor" aria-label="anchor" href="#limma-analysis"></a>
</h2>
<p>Limma has a built-in approach for analyzing repeated measures data
using <code><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">duplicateCorrelation()</a></code>. The model can handle a single
random effect, and forces the magnitude of the random effect to be the
same across all genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># apply duplicateCorrelation is two rounds</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">Disease</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">vobj_tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">dupcor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">duplicateCorrelation</a></span><span class="op">(</span><span class="va">vobj_tmp</span>, <span class="va">design</span>, block <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run voom considering the duplicateCorrelation results</span></span>
<span><span class="co"># in order to compute more accurate precision weights</span></span>
<span><span class="co"># Otherwise, use the results from the first voom run</span></span>
<span><span class="va">vobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span>, plot <span class="op">=</span> <span class="cn">FALSE</span>, block <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span>, correlation <span class="op">=</span> <span class="va">dupcor</span><span class="op">$</span><span class="va">consensus</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate linear mixed model with a single variance component</span></span>
<span><span class="co"># Fit the model for each gene,</span></span>
<span><span class="va">dupcor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">duplicateCorrelation</a></span><span class="op">(</span><span class="va">vobj</span>, <span class="va">design</span>, block <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># But this step uses only the genome-wide average for the random effect</span></span>
<span><span class="va">fitDupCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">vobj</span>, <span class="va">design</span>, block <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span>, correlation <span class="op">=</span> <span class="va">dupcor</span><span class="op">$</span><span class="va">consensus</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit Empirical Bayes for moderated t-statistics</span></span>
<span><span class="va">fitDupCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fitDupCor</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dream-analysis">Dream Analysis<a class="anchor" aria-label="anchor" href="#dream-analysis"></a>
</h2>
<p>The dream method replaces 4 core functions of limma with a linear
mixed model.</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code> replaces <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom()</a></code> to
estimate precision weights</li>
<li>
<code><a href="../reference/dream-method.html">dream()</a></code> replaces <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit()</a></code> to estimate
regression coefficients.<br>
</li>
<li>
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">variancePartition::eBayes()</a></code> replaces
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">limma::eBayes()</a></code> to apply empircal Bayes shrinkage on linear
mixed models.</li>
<li>
<code><a href="../reference/toptable-method.html">variancePartition::topTable()</a></code> replaces
<code><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">limma::topTable()</a></code> to give seamless access to results from
<code><a href="../reference/dream-method.html">dream()</a></code>.</li>
</ol>
<p>For models with only fixed effects,
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">variancePartition::eBayes()</a></code>, and
<code><a href="../reference/toptable-method.html">variancePartition::topTable()</a></code> work seamlessly and give
results equivalent to the <code>limma</code> functions with the same
name. From the user perspective, the <code><a href="../reference/dream-method.html">dream()</a></code> workflow is
the same as <code>limma</code> since the statistical differences are
handled behind the scenes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify parallel processing parameters</span></span>
<span><span class="co"># this is used implicitly by dream() to run in parallel</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"SOCK"</span>, progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The variable to be tested must be a fixed effect</span></span>
<span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Disease</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Individual</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># estimate weights using linear mixed model of dream</span></span>
<span><span class="va">vobjDream</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">form</span>, <span class="va">metadata</span>, BPPARAM <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit the dream model on each gene</span></span>
<span><span class="co"># For the hypothesis testing, by default,</span></span>
<span><span class="co"># dream() uses the KR method for &lt;= 20 samples,</span></span>
<span><span class="co"># otherwise it uses the Satterthwaite approximation</span></span>
<span><span class="va">fitmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">fitmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fitmm</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Examine design matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fitmm</span><span class="op">$</span><span class="va">design</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           (Intercept) Disease1</span></span>
<span><span class="co">## sample_01           1        0</span></span>
<span><span class="co">## sample_02           1        0</span></span>
<span><span class="co">## sample_03           1        0</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get results of hypothesis test on coefficients of interest</span></span>
<span><span class="fu"><a href="../reference/toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">fitmm</span>, coef <span class="op">=</span> <span class="st">"Disease1"</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                   logFC  AveExpr        t      P.Value    adj.P.Val        B</span></span>
<span><span class="co">## ENST00000283033.5 gene=TXNDC11 1.556233 3.567624 32.66731 1.934507e-21 1.934507e-18 38.78853</span></span>
<span><span class="co">## ENST00000257181.9 gene=PRPF38A 1.380549 4.398270 24.57867 1.493752e-18 7.468760e-16 32.48845</span></span>
<span><span class="co">## ENST00000525790.1 gene=TDRKH   1.508341 3.184931 19.32367 3.712012e-16 1.025502e-13 27.06758</span></span>
<span><span class="co">##                                   z.std</span></span>
<span><span class="co">## ENST00000283033.5 gene=TXNDC11 9.508490</span></span>
<span><span class="co">## ENST00000257181.9 gene=PRPF38A 8.790139</span></span>
<span><span class="co">## ENST00000525790.1 gene=TDRKH   8.147605</span></span></code></pre>
<p>Since dream uses an estimated degrees of freedom value for each
hypothsis test, the degrees of freedom is different for each gene here.
Therefore, the t-statistics are not directly comparable since they have
different degrees of freedom. In order to be able to compare test
statistics, we report <code>z.std</code> which is the p-value
transformed into a signed z-score. This can be used for downstream
analysis.</p>
<p>Note that if a random effect is not specified, <code><a href="../reference/dream-method.html">dream()</a></code>
automatically uses <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit()</a></code>, but the user must run
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes()</a></code> afterward.</p>
<div class="section level3">
<h3 id="advanced-hypothesis-testing">Advanced hypothesis testing<a class="anchor" aria-label="anchor" href="#advanced-hypothesis-testing"></a>
</h3>
<div class="section level4">
<h4 id="using-contrasts-to-compare-coefficients">Using contrasts to compare coefficients<a class="anchor" aria-label="anchor" href="#using-contrasts-to-compare-coefficients"></a>
</h4>
<p>You can also perform a hypothesis test of the <em>difference</em>
between two or more coefficients by using a contrast matrix. The
contrasts are evaluated at the time of the model fit and the results can
be extracted with <code><a href="../reference/toptable-method.html">topTable()</a></code>. This behaves like
<code><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts()</a></code> and <code><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit()</a></code> in
limma.</p>
<p>Multiple contrasts can be evaluated at the same time, in order to
save computation time. Make sure to inspect your contrast matrix to
confirm it is testing what you intend.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">DiseaseSubtype</span> <span class="op">+</span> <span class="va">Sex</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Individual</span><span class="op">)</span></span>
<span></span>
<span><span class="va">L</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeContrastsDream.html">makeContrastsDream</a></span><span class="op">(</span><span class="va">form</span>, <span class="va">metadata</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    compare2_1 <span class="op">=</span> <span class="st">"DiseaseSubtype2 - DiseaseSubtype1"</span>,</span>
<span>    compare1_0 <span class="op">=</span> <span class="st">"DiseaseSubtype1 - DiseaseSubtype0"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize contrast matrix</span></span>
<span><span class="fu"><a href="../reference/plotContrasts.html">plotContrasts</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span></span></code></pre></div>
<p><img src="dream_files/figure-html/contrast-1.png" width="768"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fit dream model with contrasts</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, <span class="va">L</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get names of available coefficients and contrasts for testing</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "compare2_1"      "compare1_0"      "DiseaseSubtype0" "DiseaseSubtype1" "DiseaseSubtype2"</span></span>
<span><span class="co">## [6] "SexM"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract results from first contrast</span></span>
<span><span class="fu"><a href="../reference/toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"compare2_1"</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                       logFC  AveExpr         t      P.Value adj.P.Val          B</span></span>
<span><span class="co">## ENST00000355624.3 gene=RAB11FIP2 -0.9493146 5.260280 -5.350356 0.0000224728 0.0224728  0.8858852</span></span>
<span><span class="co">## ENST00000593466.1 gene=DDA1      -1.7265710 3.901579 -3.664795 0.0013569332 0.6784666 -1.5402278</span></span>
<span><span class="co">## ENST00000200676.3 gene=CETP       1.4777422 3.723438  3.873141 0.0040308941 0.9994857 -1.8241804</span></span>
<span><span class="co">##                                      z.std</span></span>
<span><span class="co">## ENST00000355624.3 gene=RAB11FIP2 -4.238790</span></span>
<span><span class="co">## ENST00000593466.1 gene=DDA1      -3.203659</span></span>
<span><span class="co">## ENST00000200676.3 gene=CETP       2.875734</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="comparing-multiple-coefficients">Comparing multiple coefficients<a class="anchor" aria-label="anchor" href="#comparing-multiple-coefficients"></a>
</h4>
<p>So far contrasts have only involved the difference between two
coefficients. But contrasts can also compare any linear combination of
coefficients. Here, consider comparing <code>DiseaseSubtype0</code> to
the mean of <code>DiseaseSubtype1</code> and
<code>DiseaseSubtype2</code>. Note you can also customize the name of
the contrast.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">L2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeContrastsDream.html">makeContrastsDream</a></span><span class="op">(</span><span class="va">form</span>, <span class="va">metadata</span>,</span>
<span>  contrasts <span class="op">=</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Test1 <span class="op">=</span> <span class="st">"DiseaseSubtype0 - (DiseaseSubtype1 + DiseaseSubtype2)/2"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotContrasts.html">plotContrasts</a></span><span class="op">(</span><span class="va">L2</span><span class="op">)</span></span></code></pre></div>
<p><img src="dream_files/figure-html/maual.contrasts-1.png" width="768"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fit dream model to evaluate contrasts</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobjDream</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span>, <span class="va">form</span>, <span class="va">metadata</span>, L <span class="op">=</span> <span class="va">L2</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"Test1"</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                     logFC  AveExpr         t      P.Value    adj.P.Val        B</span></span>
<span><span class="co">## ENST00000418210.2 gene=TMEM64  -1.0343236 4.715367 -7.000080 3.300023e-07 3.054320e-06 8.484489</span></span>
<span><span class="co">## ENST00000456159.1 gene=MET     -0.9830788 2.458926 -6.062227 6.108639e-07 3.054320e-06 5.836340</span></span>
<span><span class="co">## ENST00000555834.1 gene=RPS6KL1 -0.8954795 5.272063 -5.450341 3.957534e-06 1.319178e-05 3.997352</span></span>
<span><span class="co">##                                    z.std</span></span>
<span><span class="co">## ENST00000418210.2 gene=TMEM64  -5.105453</span></span>
<span><span class="co">## ENST00000456159.1 gene=MET     -4.987750</span></span>
<span><span class="co">## ENST00000555834.1 gene=RPS6KL1 -4.613600</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="joint-hypothesis-test-of-multiple-coefficients">Joint hypothesis test of multiple coefficients<a class="anchor" aria-label="anchor" href="#joint-hypothesis-test-of-multiple-coefficients"></a>
</h4>
<p>Joint hypothesis testing of multiple coefficients at the same time
can be performed by using an F-test. Just like in limma, the results can
be extracted using <code><a href="../reference/toptable-method.html">topTable()</a></code></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract results from first contrast</span></span>
<span><span class="fu"><a href="../reference/toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DiseaseSubtype2"</span>, <span class="st">"DiseaseSubtype1"</span><span class="op">)</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                DiseaseSubtype2 DiseaseSubtype1  AveExpr        F      P.Value</span></span>
<span><span class="co">## ENST00000418210.2 gene=TMEM64         5.301001        5.211674 4.715367 823.4480 3.966577e-23</span></span>
<span><span class="co">## ENST00000555834.1 gene=RPS6KL1        5.662699        5.719196 5.272063 791.4321 6.389688e-23</span></span>
<span><span class="co">## ENST00000589123.1 gene=NFIC           6.545195        6.181023 5.855335 651.9138 6.549307e-22</span></span>
<span><span class="co">##                                   adj.P.Val    F.std</span></span>
<span><span class="co">## ENST00000418210.2 gene=TMEM64  3.194844e-22 51.58155</span></span>
<span><span class="co">## ENST00000555834.1 gene=RPS6KL1 3.194844e-22 51.10477</span></span>
<span><span class="co">## ENST00000589123.1 gene=NFIC    2.183102e-21 48.77751</span></span></code></pre>
<p>Since dream uses an estimated degrees of freedom value for each
hypothsis test, the degrees of freedom is different for each gene here.
Therefore, the F-statistics are not directly comparable since they have
different degrees of freedom. In order to be able to compare test
statistics, we report <code>F.std</code> which is the p-value
transformed into an F-statistic with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><msub><mi>f</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">df_1</annotation></semantics></math>
the number of coefficients tested and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><msub><mi>f</mi><mn>2</mn></msub><mo>=</mo><mi>∞</mi></mrow><annotation encoding="application/x-tex">df_2=\infty</annotation></semantics></math>.
This can be used for downstream analysis.</p>
</div>
</div>
<div class="section level3">
<h3 id="small-sample-method">Small-sample method<a class="anchor" aria-label="anchor" href="#small-sample-method"></a>
</h3>
<p>For small datasets, the Kenward-Roger method can be more powerful.
But it is <strong>substantially</strong> more computationally
intensive.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fitmmKR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, ddf <span class="op">=</span> <span class="st">"Kenward-Roger"</span><span class="op">)</span></span>
<span><span class="va">fitmmKR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fitmmKR</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="variancepartition-plot">variancePartition plot<a class="anchor" aria-label="anchor" href="#variancepartition-plot"></a>
</h3>
<p>Dream and variancePartition share the same underlying linear mixed
model framework. A variancePartition analysis can indicate important
variables that should be included as fixed or random effects in the
dream analysis.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note: this could be run with either vobj from voom()</span></span>
<span><span class="co"># or vobjDream from voomWithDreamWeights()</span></span>
<span><span class="co"># The resuylts are similar</span></span>
<span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Individual</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Disease</span><span class="op">)</span></span>
<span><span class="va">vp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitExtractVarPartModel-method.html">fitExtractVarPartModel</a></span><span class="op">(</span><span class="va">vobj</span>, <span class="va">form</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotVarPart-method.html">plotVarPart</a></span><span class="op">(</span><span class="fu"><a href="../reference/sortCols-method.html">sortCols</a></span><span class="op">(</span><span class="va">vp</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="dream_files/figure-html/vp-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="comparing-p-values">Comparing p-values<a class="anchor" aria-label="anchor" href="#comparing-p-values"></a>
</h3>
<p>Here we compare p-values from from <code><a href="../reference/dream-method.html">dream()</a></code> and
<code>duplicateCorrelation</code>. In order to understand the empircal
difference between them, we can plot the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>−</mi><msub><mo>log</mo><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-\log_{10}</annotation></semantics></math>
p-values from each method.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare p-values and make plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">fitDupCor</span>, coef <span class="op">=</span> <span class="st">"Disease1"</span>, number <span class="op">=</span> <span class="cn">Inf</span>, sort.by <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span><span class="op">$</span><span class="va">P.Value</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">fitmm</span>, number <span class="op">=</span> <span class="cn">Inf</span>, sort.by <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span><span class="op">$</span><span class="va">P.Value</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotCompareP-method.html">plotCompareP</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">vp</span><span class="op">$</span><span class="va">Individual</span>, <span class="va">dupcor</span><span class="op">$</span><span class="va">consensus</span><span class="op">)</span></span></code></pre></div>
<p><img src="dream_files/figure-html/define-1.png" width="700"></p>
<p>The duplicateCorrelation method estimates a single variance term
genome-wide even though the donor contribution of a particular gene can
vary substantially from the genome-wide trend. Using a single value
genome-wide for the within-donor variance can reduce power and increase
the false positive rate in a particular, reproducible way. Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\tau^2_g</annotation></semantics></math>
be the value of the donor component for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup><annotation encoding="application/x-tex">\bar{\tau}^2</annotation></semantics></math>
be the genome-wide mean. For genes where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><mo>&gt;</mo><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">\tau^2_g&gt;\bar{\tau}^2</annotation></semantics></math>,
using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup><annotation encoding="application/x-tex">\bar{\tau}^2</annotation></semantics></math>
under-corrects for the donor component so that it increases the false
positive rate compared to using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\tau^2_g</annotation></semantics></math>.
Conversely, for genes where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><mo>&lt;</mo><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">\tau^2_g&lt;\bar{\tau}^2</annotation></semantics></math>,
using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup><annotation encoding="application/x-tex">\bar{\tau}^2</annotation></semantics></math>
over-corrects for the donor component so that it decreases power.
Increasing sample size does not overcome this issue. The dream method
overcomes this issue by using a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\tau^2_g</annotation></semantics></math>.</p>
<p>Here, the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>−</mi><msub><mo>log</mo><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-\log_{10}</annotation></semantics></math>
p-values from both methods are plotted and colored by the donor
contribution estiamted by variancePartition. The green value indicates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup><annotation encoding="application/x-tex">\bar{\tau}^2</annotation></semantics></math>,
while red and blue indicate higher and lower values, respectively. When
only one variance component is used and the contrast matrix is simple,
the effect of using dream versus duplicateCorrelation is determined by
the comparison of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\tau^2_g</annotation></semantics></math>
to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup><annotation encoding="application/x-tex">\bar{\tau}^2</annotation></semantics></math>:</p>
<p>dream can increase the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>−</mi><msub><mo>log</mo><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-\log_{10}</annotation></semantics></math>
p-value for genes with a lower donor component
(i.e. <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><mo>&lt;</mo><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">\tau^2_g&lt;\bar{\tau}^2</annotation></semantics></math>)
and decrease
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>−</mi><msub><mo>log</mo><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-\log_{10}</annotation></semantics></math>
p-value for genes with a higher donor component
(i.e. <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>τ</mi><mi>g</mi><mn>2</mn></msubsup><mo>&gt;</mo><msup><mover><mi>τ</mi><mo accent="true">‾</mo></mover><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">\tau^2_g&gt;\bar{\tau}^2</annotation></semantics></math>)</p>
<p>Note that using more variance components or a more complicated
contrast matrix can make the relationship more complicated.</p>
</div>
</div>
<div class="section level2">
<h2 id="parallel-processing">Parallel processing<a class="anchor" aria-label="anchor" href="#parallel-processing"></a>
</h2>
<p>variancePartition functions including <code><a href="../reference/dream-method.html">dream()</a></code>,
<code><a href="../reference/fitExtractVarPartModel-method.html">fitExtractVarPartModel()</a></code> and <code><a href="../reference/fitVarPartModel-method.html">fitVarPartModel()</a></code>
can take advange of multicore machines to speed up analysis. It uses the
<a href="http://bioconductor.org/packages/BiocParallel/" class="external-link">BiocParallel</a>
package to manage the parallelization.</p>
<ul>
<li>Specify parameters with the BPPARAM argument.</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request 4 cores, and enable the progress bar</span></span>
<span><span class="co"># This is the ideal for Linux, OS X and Windows</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"SOCK"</span>, progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">fitmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, BPPARAM <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span></code></pre></div>
<p>By default <code>BPPARAM = SerialParam()</code>.</p>
<!---
By default `BPPARAM` and the global setttings are set the results of `bpparam()`.  But note that using `SnowParam()` can *dramatically* reduce the memory usage needed for parallel processing because it reduces memory redundancy between threads.
--->
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin22.4.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS 14.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Users/gabrielhoffman/prog/R-4.3.0/lib/libRblas.dylib </span></span>
<span><span class="co">## LAPACK: /usr/local/Cellar/r/4.4.0_1/lib/R/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] knitr_1.43                edgeR_3.42.4              variancePartition_1.33.15</span></span>
<span><span class="co">## [4] BiocParallel_1.34.2       limma_3.56.2              ggplot2_3.4.4            </span></span>
<span><span class="co">## [7] pander_0.6.5             </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.0    farver_2.1.1        dplyr_1.1.2         bitops_1.0-7       </span></span>
<span><span class="co">##  [5] fastmap_1.1.1       digest_0.6.33       lifecycle_1.0.3     statmod_1.5.0      </span></span>
<span><span class="co">##  [9] magrittr_2.0.3      compiler_4.3.0      rlang_1.1.1         sass_0.4.9         </span></span>
<span><span class="co">## [13] tools_4.3.0         utf8_1.2.3          yaml_2.3.7          labeling_0.4.2     </span></span>
<span><span class="co">## [17] htmlwidgets_1.6.2   plyr_1.8.8          KernSmooth_2.23-21  withr_2.5.0        </span></span>
<span><span class="co">## [21] purrr_1.0.2         numDeriv_2016.8-1.1 BiocGenerics_0.46.0 desc_1.4.2         </span></span>
<span><span class="co">## [25] grid_4.3.0          aod_1.3.2           fansi_1.0.4         caTools_1.18.2     </span></span>
<span><span class="co">## [29] colorspace_2.1-0    scales_1.2.1        gtools_3.9.4        iterators_1.0.14   </span></span>
<span><span class="co">## [33] MASS_7.3-60         cli_3.6.1           mvtnorm_1.2-2       rmarkdown_2.22     </span></span>
<span><span class="co">## [37] ragg_1.2.5          generics_0.1.3      reshape2_1.4.4      minqa_1.2.5        </span></span>
<span><span class="co">## [41] cachem_1.0.8        stringr_1.5.0       splines_4.3.0       parallel_4.3.0     </span></span>
<span><span class="co">## [45] matrixStats_1.0.0   vctrs_0.6.3         boot_1.3-28.1       Matrix_1.6-5       </span></span>
<span><span class="co">## [49] jsonlite_1.8.5      pbkrtest_0.5.2      systemfonts_1.0.4   locfit_1.5-9.7     </span></span>
<span><span class="co">## [53] jquerylib_0.1.4     tidyr_1.3.1         snow_0.4-4          glue_1.6.2         </span></span>
<span><span class="co">## [57] pkgdown_2.0.9       nloptr_2.0.3        codetools_0.2-19    stringi_1.7.12     </span></span>
<span><span class="co">## [61] gtable_0.3.3        EnvStats_2.7.0      lme4_1.1-35.1       lmerTest_3.1-3     </span></span>
<span><span class="co">## [65] munsell_0.5.0       tibble_3.2.1        remaCor_0.0.18      pillar_1.9.0       </span></span>
<span><span class="co">## [69] htmltools_0.5.8.1   gplots_3.1.3        R6_2.5.1            textshaping_0.3.6  </span></span>
<span><span class="co">## [73] Rdpack_2.4          rprojroot_2.0.3     evaluate_0.21       lattice_0.21-8     </span></span>
<span><span class="co">## [77] Biobase_2.60.0      highr_0.10          rbibutils_2.2.13    backports_1.4.1    </span></span>
<span><span class="co">## [81] RhpcBLASctl_0.23-42 memoise_2.0.1       broom_1.0.5         fANCOVA_0.6-1      </span></span>
<span><span class="co">## [85] corpcor_1.6.10      bslib_0.7.0         Rcpp_1.0.11         nlme_3.1-162       </span></span>
<span><span class="co">## [89] xfun_0.39           fs_1.6.2            pkgconfig_2.0.3</span></span></code></pre>
<p>&lt;&gt;</p>
</details>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Law2014" class="csl-entry">
Law, C. W., Y. Chen, W. Shi, and G. K. Smyth. 2014. <span>“<span class="nocase">Voom: precision weights unlock linear model analysis
tools for RNA-seq read counts.</span>”</span> <em>Genome Biology</em> 15
(2): R29. <a href="https://doi.org/10.1186/gb-2014-15-2-r29" class="external-link">https://doi.org/10.1186/gb-2014-15-2-r29</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
