<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multivariate tests â€¢ variancePartition</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multivariate tests">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.39.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/variancePartition.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/dream.html">dream analysis</a></li>
    <li><a class="dropdown-item" href="../articles/rnd_effects.html">Theory and practice of random effects</a></li>
    <li><a class="dropdown-item" href="../articles/mvtests.html">Multivariate tests</a></li>
    <li><a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Multivariate tests</h1>
            <h3 data-toc-skip class="subtitle">Combining results of
univariate tests</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2025-08-19
20:57:20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/vignettes/mvtests.Rmd" class="external-link"><code>vignettes/mvtests.Rmd</code></a></small>
      <div class="d-none name"><code>mvtests.Rmd</code></div>
    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes

rmarkdown::render('multivariate_tests.Rmd')

--->
<style>
body {
text-align: justify}
</style>
<p>Results from the univariate regressions performed using can be
combined in a post-processing step to perform multivariate hypothesis
testing. In this example, we fit on transcript-level counts and then
perform multivariate hypothesis testing by combining transcripts at the
gene-level. This is done with the function.</p>
<div class="section level2">
<h2 id="import-transcript-level-counts">Import transcript-level counts<a class="anchor" aria-label="anchor" href="#import-transcript-level-counts"></a>
</h2>
<p>Read in transcript counts from the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/tximport" class="external-link">tximport</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tximportData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify directory</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"tximportData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read sample meta-data</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"samples.txt"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">samples.ext</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"samples_extended.txt"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read assignment of transcripts to genes</span></span>
<span><span class="co"># remove genes on the PAR, since these are present twice</span></span>
<span><span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"tx2gene.gencode.v27.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="va">tx2gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"PAR_Y"</span>, <span class="va">tx2gene</span><span class="op">$</span><span class="va">GENEID</span>, invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># read transcript-level quatifictions</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"salmon"</span>, <span class="va">samples</span><span class="op">$</span><span class="va">run</span>, <span class="st">"quant.sf.gz"</span><span class="op">)</span></span>
<span><span class="va">txi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport</a></span><span class="op">(</span><span class="va">files</span>, type <span class="op">=</span> <span class="st">"salmon"</span>, txOut <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create metadata simulating two conditions</span></span>
<span><span class="va">sampleTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sampleTable</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="standard-dream-analysis">Standard dream analysis<a class="anchor" aria-label="anchor" href="#standard-dream-analysis"></a>
</h2>
<p>Perform standard analysis at the transcript-level</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare transcript-level reads</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="va">txi</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, data <span class="op">=</span> <span class="va">sampleTable</span><span class="op">)</span></span>
<span><span class="va">isexpr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="va">dge</span><span class="op">[</span><span class="va">isexpr</span>, <span class="op">]</span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate precision weights</span></span>
<span><span class="va">vobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights</a></span><span class="op">(</span><span class="va">dge</span>, <span class="op">~</span><span class="va">condition</span>, <span class="va">sampleTable</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit regression model one transcript at a time</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobj</span>, <span class="op">~</span><span class="va">condition</span>, <span class="va">sampleTable</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/eBayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="multivariate-analysis">Multivariate analysis<a class="anchor" aria-label="anchor" href="#multivariate-analysis"></a>
</h2>
<p>Combine the transcript-level results at the gene-level. The mapping
between transcript and gene is stored in as a list.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare transcript to gene mapping</span></span>
<span><span class="co"># keep only transcripts present in vobj</span></span>
<span><span class="co"># then convert to list with key GENEID and values TXNAMEs</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">TXNAME</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vobj</span><span class="op">)</span></span>
<span><span class="va">tx2gene.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/stack.html" class="external-link">unstack</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run multivariate test on entries in each feature set</span></span>
<span><span class="co"># Default method is "FE.empirical", but use "FE" here to reduce runtime</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mvTest-method.html">mvTest</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">vobj</span>, <span class="va">tx2gene.lst</span>, coef <span class="op">=</span> <span class="st">"conditionB"</span>, method <span class="op">=</span> <span class="st">"FE"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># truncate gene names since they have version numbers</span></span>
<span><span class="co"># ENST00000498289.5 -&gt; ENST00000498289</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">ID.short</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..+"</span>, <span class="st">""</span>, <span class="va">res</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="gene-set-analysis">Gene set analysis<a class="anchor" aria-label="anchor" href="#gene-set-analysis"></a>
</h2>
<p>Perform gene set analysis using on the gene-level test
statistics.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># must have zenith &gt; v1.0.2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeuroGenomics.github.io/zenith" class="external-link">zenith</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/get_MSigDB.html" class="external-link">get_MSigDB</a></span><span class="op">(</span><span class="st">"C1"</span>, to <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df_gsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/zenithPR_gsa.html" class="external-link">zenithPR_gsa</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">stat</span>, <span class="va">res</span><span class="op">$</span><span class="va">ID.short</span>, <span class="va">gs</span>, inter.gene.cor <span class="op">=</span> <span class="fl">.05</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_gsa</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                NGenes Correlation      delta        se     p.less   p.greater     PValue Direction</span></span>
<span><span class="co">## M7078_chr2p16      30        0.05  1.4208384 0.5610910 0.99432899 0.005671015 0.01134203        Up</span></span>
<span><span class="co">## M14982_chr7p13     26        0.05  1.1335492 0.5777005 0.97512013 0.024879873 0.04975975        Up</span></span>
<span><span class="co">## M7314_chr4p14      25        0.05 -1.1344103 0.5825608 0.02575932 0.974240679 0.05151864      Down</span></span>
<span><span class="co">## M5824_chr11p13     30        0.05 -1.0120371 0.5612285 0.03568377 0.964316230 0.07136754      Down</span></span>
<span><span class="co">## M3783_chr2q37      73        0.05  0.8367603 0.4929617 0.95518099 0.044819012 0.08963802        Up</span></span>
<span><span class="co">## M10517_chr4q24     21        0.05 -1.0062435 0.6060832 0.04844305 0.951556955 0.09688609      Down</span></span>
<span><span class="co">##                      FDR</span></span>
<span><span class="co">## M7078_chr2p16  0.9992274</span></span>
<span><span class="co">## M14982_chr7p13 0.9992274</span></span>
<span><span class="co">## M7314_chr4p14  0.9992274</span></span>
<span><span class="co">## M5824_chr11p13 0.9992274</span></span>
<span><span class="co">## M3783_chr2q37  0.9992274</span></span>
<span><span class="co">## M10517_chr4q24 0.9992274</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin22.4.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS 14.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Users/gabrielhoffman/prog/R-4.3.0/lib/libRblas.dylib </span></span>
<span><span class="co">## LAPACK: /usr/local/Cellar/r/4.3.0_1/lib/R/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] org.Hs.eg.db_3.17.0       msigdbr_7.5.1             GSEABase_1.62.0          </span></span>
<span><span class="co">##  [4] graph_1.78.0              annotate_1.78.0           XML_3.99-0.14            </span></span>
<span><span class="co">##  [7] AnnotationDbi_1.62.1      IRanges_2.34.1            S4Vectors_0.38.1         </span></span>
<span><span class="co">## [10] Biobase_2.60.0            BiocGenerics_0.46.0       zenith_1.4.1             </span></span>
<span><span class="co">## [13] edgeR_3.42.4              variancePartition_1.33.11 BiocParallel_1.34.2      </span></span>
<span><span class="co">## [16] limma_3.56.2              ggplot2_3.4.4             tximportData_1.28.0      </span></span>
<span><span class="co">## [19] tximport_1.28.0           readr_2.1.4              </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.5              magrittr_2.0.3              nloptr_2.0.3               </span></span>
<span><span class="co">##   [4] rmarkdown_2.22              fs_1.6.2                    zlibbioc_1.46.0            </span></span>
<span><span class="co">##   [7] ragg_1.2.5                  vctrs_0.6.3                 memoise_2.0.1              </span></span>
<span><span class="co">##  [10] minqa_1.2.5                 RCurl_1.98-1.12             progress_1.2.2             </span></span>
<span><span class="co">##  [13] htmltools_0.5.5             S4Arrays_1.2.0              curl_5.0.0                 </span></span>
<span><span class="co">##  [16] broom_1.0.5                 sass_0.4.6                  KernSmooth_2.23-21         </span></span>
<span><span class="co">##  [19] bslib_0.4.2                 desc_1.4.2                  pbkrtest_0.5.2             </span></span>
<span><span class="co">##  [22] plyr_1.8.8                  cachem_1.0.8                lifecycle_1.0.3            </span></span>
<span><span class="co">##  [25] iterators_1.0.14            pkgconfig_2.0.3             Matrix_1.5-4.1             </span></span>
<span><span class="co">##  [28] R6_2.5.1                    fastmap_1.1.1               GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">##  [31] rbibutils_2.2.13            MatrixGenerics_1.12.0       digest_0.6.33              </span></span>
<span><span class="co">##  [34] numDeriv_2016.8-1.1         colorspace_2.1-0            rprojroot_2.0.3            </span></span>
<span><span class="co">##  [37] textshaping_0.3.6           GenomicRanges_1.52.0        RSQLite_2.3.1              </span></span>
<span><span class="co">##  [40] filelock_1.0.2              RcppZiggurat_0.1.6          fansi_1.0.4                </span></span>
<span><span class="co">##  [43] httr_1.4.6                  abind_1.4-5                 compiler_4.3.0             </span></span>
<span><span class="co">##  [46] bit64_4.0.5                 aod_1.3.2                   withr_2.5.0                </span></span>
<span><span class="co">##  [49] backports_1.4.1             DBI_1.1.3                   gplots_3.1.3               </span></span>
<span><span class="co">##  [52] MASS_7.3-60                 DelayedArray_0.26.3         corpcor_1.6.10             </span></span>
<span><span class="co">##  [55] gtools_3.9.4                caTools_1.18.2              tools_4.3.0                </span></span>
<span><span class="co">##  [58] remaCor_0.0.17              glue_1.6.2                  nlme_3.1-162               </span></span>
<span><span class="co">##  [61] grid_4.3.0                  reshape2_1.4.4              generics_0.1.3             </span></span>
<span><span class="co">##  [64] gtable_0.3.3                tzdb_0.4.0                  tidyr_1.3.0                </span></span>
<span><span class="co">##  [67] hms_1.1.3                   utf8_1.2.3                  XVector_0.40.0             </span></span>
<span><span class="co">##  [70] pillar_1.9.0                stringr_1.5.0               babelgene_22.9             </span></span>
<span><span class="co">##  [73] vroom_1.6.3                 splines_4.3.0               dplyr_1.1.2                </span></span>
<span><span class="co">##  [76] BiocFileCache_2.8.0         lattice_0.21-8              bit_4.0.5                  </span></span>
<span><span class="co">##  [79] tidyselect_1.2.0            locfit_1.5-9.7              Biostrings_2.68.1          </span></span>
<span><span class="co">##  [82] knitr_1.43                  SummarizedExperiment_1.30.1 RhpcBLASctl_0.23-42        </span></span>
<span><span class="co">##  [85] xfun_0.39                   matrixStats_1.0.0           KEGGgraph_1.60.0           </span></span>
<span><span class="co">##  [88] stringi_1.7.12              yaml_2.3.7                  boot_1.3-28.1              </span></span>
<span><span class="co">##  [91] evaluate_0.21               codetools_0.2-19            tibble_3.2.1               </span></span>
<span><span class="co">##  [94] Rgraphviz_2.44.0            cli_3.6.1                   xtable_1.8-4               </span></span>
<span><span class="co">##  [97] systemfonts_1.0.4           Rdpack_2.4                  munsell_0.5.0              </span></span>
<span><span class="co">## [100] jquerylib_0.1.4             Rcpp_1.0.11                 GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [103] EnvStats_2.7.0              dbplyr_2.3.2                png_0.1-8                  </span></span>
<span><span class="co">## [106] Rfast_2.0.7                 parallel_4.3.0              pkgdown_2.0.7              </span></span>
<span><span class="co">## [109] blob_1.2.4                  prettyunits_1.1.1           bitops_1.0-7               </span></span>
<span><span class="co">## [112] lme4_1.1-34                 mvtnorm_1.2-2               lmerTest_3.1-3             </span></span>
<span><span class="co">## [115] scales_1.2.1                purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">## [118] fANCOVA_0.6-1               rlang_1.1.1                 EnrichmentBrowser_2.30.1   </span></span>
<span><span class="co">## [121] KEGGREST_1.40.0</span></span></code></pre>
<p>&lt;&gt;</p>
</details>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
