<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Theory and practice of random effects â€¢ variancePartition</title>
<!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Theory and practice of random effects">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.39.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/variancePartition.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/dream.html">dream analysis</a></li>
    <li><a class="dropdown-item" href="../articles/rnd_effects.html">Theory and practice of random effects</a></li>
    <li><a class="dropdown-item" href="../articles/mvtests.html">Multivariate tests</a></li>
    <li><a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Theory and practice of random effects</h1>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2025-10-29
14:10:26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/vignettes/rnd_effects.Rmd" class="external-link"><code>vignettes/rnd_effects.Rmd</code></a></small>
      <div class="d-none name"><code>rnd_effects.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The distinction between modeling a variable as a fixed versus a
      random effect depends on the goal of the statistical analysis.
      While some theory and software make a strong distinction,
      <code>variancePartition</code> and <code>dream</code> take
      different approaches based on the goal of each type of analysis.
      Here we consider the distinction between fixed and random effects,
      and the usage of REML in <code>variancePartition</code> and
      <code>dream</code>.</p>
    </div>
    
<style>
body {
text-align: justify}
</style>
<div class="section level3">
<h3 id="variancepartition">
<code>variancePartition</code><a class="anchor" aria-label="anchor" href="#variancepartition"></a>
</h3>
<div class="section level4">
<h4 id="estimating-contributions-to-expression-variation">Estimating contributions to expression variation<a class="anchor" aria-label="anchor" href="#estimating-contributions-to-expression-variation"></a>
</h4>
<p>In traditional statistics and biostatistics, there is a strong
distinction between modeling categorical variants as fixed and random
effects. Random effects correspond to a sample of units from a larger
population, while fixed effects correspond to properties of specific
individuals. Random effects are typically treated as nuisance variables
and integrated out, and hypothesis testing is performed on the fixed
effect.</p>
<p>The <code>r2glmm</code> package fits into this traditional framework,
by computing the variance fractions for a given fixed effect as: <span class="math display">\begin{eqnarray}
\sigma^2_{fixed}/ \left(\sigma^2_{fixed} + \sigma^2_{error}\right)
\end{eqnarray}</span></p>
<p>Importantly, the random effects are not in the denominator. The
fraction is only determined by fixed effects and residuals.</p>
<p>In my experience in bioinformatics, this was a problem. Making such
distinctions between fixed and random effects seemed arbitrary. Variance
in a phenotype could be due to age (fixed) or to variation across
subject (random). Including all of the variables in the denominator
produced more intuitive results so that 1) the variance fractions sum to
one across all components and 2) fixed and random effects could be
interpreted on the same scale 3) fractions could be compared across
studies with different designs, 4) estimates of variance fractions were
most accurate. So in variancePartition the fractions are defined as:
<span class="math display">\begin{eqnarray}
\sigma^2_{X}/ \left(\sigma^2_{fixed} + \sigma^2_{random} +
\sigma^2_{error}\right)
\end{eqnarray}</span></p>
<p>just plugging the each variable in the numerator.</p>
<p>Thus the faction evaluated by variancePartition is different than
<code>r2glmm</code> by definition.</p>
<p>Here is some code explicitly demonstrating this difference:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">"variancePartition"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lme4/lme4/" class="external-link">"lme4"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/bcjaeger/r2glmm" class="external-link">"r2glmm"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate 1 fixed variable and 1 random variable with 3 levels</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span><span class="op">)</span>, Subject <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span>, <span class="fl">100</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># simulate variable</span></span>
<span><span class="co"># y = X\beta + Subject\alpha + \sigma^2</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">X</span> <span class="op">*</span> <span class="va">beta</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">data</span><span class="op">$</span><span class="va">Subject</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">alpha</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># fit model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmer.html" class="external-link">lmer</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">X</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Subject</span><span class="op">)</span>, <span class="va">data</span>, REML <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate variance fraction using variancePartition</span></span>
<span><span class="co"># include the total sum in the denominator</span></span>
<span><span class="va">frac</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calcVarPart-method.html">calcVarPart</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="va">frac</span></span></code></pre></div>
<pre><code>  Subject         X Residuals 
   0.4505    0.4952    0.0543 </code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the variance fraction excluding the random effect from the denominator</span></span>
<span><span class="co"># is the same as from r2glmm</span></span>
<span><span class="va">frac</span><span class="op">[[</span><span class="st">"X"</span><span class="op">]</span><span class="op">]</span> <span class="op">/</span> <span class="op">(</span><span class="va">frac</span><span class="op">[[</span><span class="st">"X"</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">frac</span><span class="op">[[</span><span class="st">"Residuals"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] 0.901</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using r2glmm</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/r2glmm/man/r2beta.html" class="external-link">r2beta</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<pre><code>  Effect   Rsq upper.CL lower.CL
1  Model 0.896    0.904    0.886
2      X 0.896    0.904    0.886</code></pre>
<p>So the formulas are different. But why require categorical variables
as random effects?</p>
<p>At practical level, categorical variables with too many levels are
problematic. Using a categorical variable with 200 categories as a fixed
effect is statistically unstable. There are so many degrees of freedom
that that variable will absorb a lot of variance even under the null.
Statistically, estimating the variance fraction for a variable with many
categories can be biased if that variable is a fixed effect. Therefore,
<code>variancePartition</code> requires all categorical variables to be
random effects. Modeling this variable as a random effect produces
unbiased estimates of variance fractions in practice. See simulations in
the Supplement (section 1.5) of <a href="https://doi.org/10.1186/s12859-016-1323-z" class="external-link">Hoffman and Schadt
(2016)</a>.</p>
<p>The distinction between fixed and random effects is important in the
formulation because it affects which variables are put in the
denominator. So choosing to model a variable as a fixed versus random
effect will definitely change the estimated fraction.</p>
<p>Yet for the <code>variancePartition</code> formulation, all variables
are in the denominator and it isn`t affected by the fixed/random
decision. Moreover, using a random effect empirically reduces the bias
of the estimated fraction.</p>
<p>Finally, why use maximum likelihood to estimate the paramters instead
of the default REML ()? Maximum likelihood fits all parameters jointly
so that it estimates the fixed and random effects together. This is
essential if we want to compare fixed and random effects later.
Conversely, REML estimates the random effects by removing the fixed
effects from the response before estimation. This implicitly removes the
fixed effects from the denominator when evaluating the variance
fraction. REML treats fixed effects as nuisance variables, while
<code>variancePartition</code> considers fixed effects to be a core part
of the analysis.</p>
<p>While REML produced unbiased estimates of the variance components,
the goal of <code>variancePartition</code> is to estimate the variance
fractions for fixed and random effects jointly. In simulations from the
Supplement (section 1.5) of <a href="https://doi.org/10.1186/s12859-016-1323-z" class="external-link">Hoffman and Schadt
(2016)</a>, REML produced biased estimates of the variance fractions
while maximum likelihood estimates are unbiased.</p>
</div>
</div>
<div class="section level3">
<h3 id="dream">
<code>dream</code><a class="anchor" aria-label="anchor" href="#dream"></a>
</h3>
<div class="section level4">
<h4 id="hypothesis-testing">Hypothesis testing<a class="anchor" aria-label="anchor" href="#hypothesis-testing"></a>
</h4>
<p>While <code>dream</code> is also based on a linear mixed model, the
goal of this analysis is to perform hypothesis testing on fixed effects.
Random effects are treated as nuisance variables to be integrated out,
and the approximate null distribution of a t- or F-statistic is
constructed from the model fit.</p>
<p>Since the goal of the analysis is different, the consideration of
using REML versus ML is different than above. While is required by
called by when , can be used with as either or . Since the Kenward-Roger
method gave the best power with an accurate control of false positive
rate in our simulations, and since the Satterthwaite method with gives
p-values that are slightly closer to the Kenward-Roger p-values, is set
as the default.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code>R version 4.5.1 (2025-06-13)
Platform: aarch64-apple-darwin23.6.0
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS/LAPACK: /opt/homebrew/Cellar/openblas/0.3.30/lib/libopenblasp-r0.3.30.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.50

loaded via a namespace (and not attached):
 [1] digest_0.6.37     desc_1.4.3        R6_2.6.1          fastmap_1.2.0    
 [5] xfun_0.53         cachem_1.1.0      htmltools_0.5.8.1 rmarkdown_2.30   
 [9] lifecycle_1.0.4   cli_3.6.5         sass_0.4.10       pkgdown_2.1.3    
[13] textshaping_1.0.4 jquerylib_0.1.4   systemfonts_1.3.1 compiler_4.5.1   
[17] tools_4.5.1       ragg_1.5.0        evaluate_1.0.5    bslib_0.9.0      
[21] yaml_2.3.10       jsonlite_2.0.0    rlang_1.1.6       fs_1.6.6         
[25] htmlwidgets_1.6.4</code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
