<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Variance partitioning analysis • variancePartition</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variance partitioning analysis"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.36.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/variancePartition.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dream.html">dream analysis</a></li>
    <li><a class="dropdown-item" href="../articles/rnd_effects.html">Theory and practice of random effects</a></li>
    <li><a class="dropdown-item" href="../articles/mvtests.html">Multivariate tests</a></li>
    <li><a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Variance partitioning analysis</h1>
                        <h4 data-toc-skip class="author">Developed by <a
href="http://gabrielhoffman.github.io/">Gabriel Hoffman</a></h4>
            
            <h4 data-toc-skip class="date">Run on 2024-11-05

15:49:44</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/vignettes/variancePartition.Rmd" class="external-link"><code>vignettes/variancePartition.Rmd</code></a></small>
      <div class="d-none name"><code>variancePartition.Rmd</code></div>
    </div>

    
    
<!---
cd /Users/gabrielhoffman/workspace/repos
R
rmarkdown::render('variancePartition.Rmd')

--->
<style>
body {
text-align: justify}
</style>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Gene expression datasets are complicated and have multiple sources of
biological and technical variation. These datasets have recently become
more complex as it is now feasible to assay gene expression from the
same individual in multiple tissues or at multiple time points. The
<code>variancePartition</code> package implements a statistical method
to quantify the contribution of multiple sources of variation and
decouple within/between-individual variation. In addition,
<code>variancePartition</code> produces results at the gene-level to
identity genes that follow or deviate from the genome-wide trend.</p>
<p>The <code>variancePartition</code> package provides a general
framework for understanding drivers of variation in gene expression in
experiments with complex designs. A typical application would consider a
dataset of gene expression from individuals sampled in multiple tissues
or multiple time points where the goal is to understand variation within
versus between individuals and tissues. <code>variancePartition</code>
use a linear mixed model to partition the variance attributable to
multiple variables in the data. The analysis is built on top of the
<code>lme4</code> package <span class="citation">(Bates et al.
2015)</span>, and some basic knowledge about linear mixed models will
give you some intuition about the behavior of
<code>variancePartition</code> <span class="citation">(Pinheiro and
Bates 2000; Galecki and Burzykowski 2013)</span>.</p>
<div id="input-data" class="section level2">
<h2>Input data</h2>
<p>There are three components to an analysis:</p>
<ol style="list-style-type: decimal">
<li><strong>Gene expression data:</strong> In general, this is a matrix
of normalized gene expression values with genes as rows and experiments
as columns.</li>
</ol>
<ul>
<li><p>Count-based quantification: <code>featureCounts</code> <span
class="citation">(Liao, Smyth, and Shi 2014)</span>, <code>HTSeq</code>
<span class="citation">(Anders, Pyl, and Huber 2015)</span></p>
<p>Counts mapping to each gene can be normalized using counts per
million (CPM), reads per kilobase per million (RPKM) or fragments per
kilobase per million (FPKM). These count results can be processed with
<code>limma::voom()</code> <span class="citation">(Law et al.
2014)</span> to model the precision of each observation or
<code>DESeq2</code> <span class="citation">(Love, Huber, and Anders
2014)</span>.</p></li>
<li><p>Isoform quantification: <code>kallisto</code> <span
class="citation">(Bray et al. 2016)</span>, <code>sailfish</code> <span
class="citation">(Patro, Mount, and Kingsford 2014)</span>,
<code>salmon</code> <span class="citation">(Patro, Duggal, and Kingsford
2015)</span>, <code>RSEM</code> <span class="citation">(Li and Dewey
2011)</span>, <code>cufflinks</code> <span class="citation">(Trapnell et
al. 2010)</span></p>
<p>These perform isoform-level quantification using reads that map to
multiple transcripts. Quantification values can be read directly into R,
or processed with <code>ballgown</code> <span class="citation">(Frazee
et al. 2015)</span> or <code>tximport</code> <span
class="citation">(Soneson, Love, and Robinson 2015)</span>.</p></li>
<li><p>Microarray data: any standard normalization such as
<code>rma</code> in the <code>oligo</code> <span
class="citation">(Carvalho and Irizarry 2010)</span> package can be
used.</p></li>
<li><p>Any set of features: chromatin accessibility, protein
quantification, etc</p></li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><p><strong>Metadata about each experiment:</strong></p>
<p>A <code>data.frame</code> with information about each experiment such
as patient ID, tissue, sex, disease state, time point, batch,
etc.</p></li>
<li><p><strong>Formula indicating which metadata variables to
consider:</strong></p>
<p>An R formula such as
<code>~ Age + (1|Individual) + (1|Tissue) + (1|Batch)</code> indicating
which metadata variables should be used in the analysis.</p></li>
</ol>
<p>Variance partitioning analysis will assess the contribution of each
metadata variable to variation in gene expression and can report the
intra-class correlation for each variable.</p>
</div>
</div>
<div id="running-an-analysis" class="section level1">
<h1>Running an analysis</h1>
<p>A typical analysis with <code>variancePartition</code> is only a few
lines of R code, assuming the expression data has already been
normalized. Normalization is a separate topic addressed briefly in
[Applying <code>variancePartition</code> to RNA-seq expression
data].</p>
<p>The simulated dataset included as an example contains measurements of
200 genes from 100 samples. These samples include assays from 3 tissues
across 25 individuals processed in 4 batches. The individuals range in
age from 36 to 73. A typical <code>variancePartition</code> analysis
will assess the contribution of each aspect of the study design
(i.e. individual, tissue, batch, age) to the expression variation of
each gene. The analysis will prioritize these axes of variation based on
a genome-wide summary and give results at the gene-level to identity
genes that follow or deviate from this genome-wide trend. The results
can be visualized using custom plots and can be used for downstream
analysis.</p>
<div id="standard-application" class="section level2">
<h2>Standard application</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># load library</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;variancePartition&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># load simulated data:</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># geneExpr: matrix of gene expression values</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># info: information/metadata about each sample</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">data</span>(varPartData)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co"># Specify variables to consider</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co"># Age is continuous so model it as a fixed effect</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co"># Individual and Tissue are both categorical,</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co"># so model them as random effects</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># Note the syntax used to specify random effects</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> Age <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch)</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co"># Fit model and extract results</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co"># 1) fit linear mixed model on gene expression</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co"># If categorical variables are specified,</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="co">#     a linear mixed model is used</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="co"># If all variables are modeled as fixed effects,</span></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co">#       a linear model is used</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="co"># each entry in results is a regression model fit on a single gene</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="co"># 2) extract variance fractions from each model fit</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="co"># for each gene, returns fraction of variation attributable</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="co">#       to each variable</span></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="co"># Interpretation: the variance explained by each variables</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="co"># after correcting for all other variables</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="co"># Note that geneExpr can either be a matrix,</span></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="co"># and EList output by voom() in the limma package,</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="co"># or an ExpressionSet</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(geneExpr, form, info)</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="co"># sort variables (i.e. columns) by median fraction</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="co">#       of variance explained</span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a>vp <span class="ot">&lt;-</span> <span class="fu">sortCols</span>(varPart)</span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="co"># Figure 1a</span></span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="co"># Bar plot of variance fractions for the first 10 genes</span></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="fu">plotPercentBars</span>(vp[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ])</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/simResult-1.png" width="384" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Figure 1b</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co"># violin plot of contribution of each variable to total variance</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">plotVarPart</span>(vp)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/simResult-2.png" width="384" style="display: block; margin: auto;" /></p>
<p><code>variancePartition</code> includes a number of custom plots to
visualize the results. Since <code>variancePartition</code> attributes
the fraction of total variation attributable to each aspect of the study
design, these fractions naturally sum to 1. plots the partitioning
results for a subset of genes (Figure 1a), and shows a genome-wide
violin plot of the distribution of variance explained by each variable
across all genes (Figure 1b). (Note that these plots show results in
terms of of variance explained, while the results are stored in terms of
the <u>fraction</u>.)</p>
<p>The core functions of <code>variancePartition</code> work seemlessly
with gene expression data stored as a <code>matrix</code>,
<code>data.frame</code>, <code>EList</code> from <code>limma</code> or
<code>ExpressionSet</code> from <code>Biobase</code>.
<code>fitExtractVarPartModel()</code> returns an object that stores the
variance fractions for each gene and each variable in the formula
specified. These fractions can be accessed just like a
<code>data.frame</code>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Access first entries</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">head</span>(varPart)</span></code></pre></div>
<pre><code>##             Batch Individual     Tissue          Age  Residuals
## gene1 0.000157945  0.8903733 0.02468893 4.528906e-05 0.08473455
## gene2 0.000000000  0.8060314 0.10101922 3.336680e-04 0.09261567
## gene3 0.002422174  0.8901096 0.03562102 1.471701e-03 0.07037550
## gene4 0.000000000  0.7688278 0.12531362 1.014413e-03 0.10484419
## gene5 0.000000000  0.6997242 0.20910139 3.871485e-05 0.09113569
## gene6 0.002343666  0.7222283 0.16786542 2.717378e-03 0.10484521</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Access first entries for Individual</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">head</span>(varPart<span class="sc">$</span>Individual)</span></code></pre></div>
<pre><code>## [1] 0.8903733 0.8060314 0.8901096 0.7688278 0.6997242 0.7222283</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># sort genes based on variance explained by Individual</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">head</span>(varPart[<span class="fu">order</span>(varPart<span class="sc">$</span>Individual, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ])</span></code></pre></div>
<pre><code>##               Batch Individual      Tissue          Age  Residuals
## gene43  0.000000000  0.9143523 0.011737153 3.776742e-04 0.07353289
## gene174 0.000000000  0.9112702 0.009726001 2.015102e-03 0.07698865
## gene111 0.000000000  0.9067082 0.008386574 9.735963e-04 0.08393165
## gene127 0.000000000  0.9035627 0.013838828 5.081822e-04 0.08209028
## gene151 0.006079206  0.9029393 0.000000000 1.345019e-05 0.09096803
## gene91  0.000000000  0.9002946 0.014144910 1.110514e-06 0.08555939</code></pre>
<div id="saving-plot-to-file" class="section level4">
<h4>Saving plot to file</h4>
<p>In order to save the plot to a file, use the <code>ggsave()</code>
function:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>fig <span class="ot">&lt;-</span> <span class="fu">plotVarPart</span>(vp)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">ggsave</span>(file, fig)</span></code></pre></div>
</div>
<div id="plot-expression-stratified-by-other-variables"
class="section level3">
<h3>Plot expression stratified by other variables</h3>
<p><code>variancePartition</code> also includes plotting functions to
visualize the variation across a variable of interest. plots the
expression of a gene stratified by the specified variable. In the
example dataset, users can plot a gene expression trait stratified by
Tissue (Figure 2a) or Individual (Figure 2b).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># get gene with the highest variation across Tissues</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co"># create data.frame with expression of gene i and Tissue</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co">#       type for each sample</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="fu">which.max</span>(varPart<span class="sc">$</span>Tissue)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>GE <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Expression =</span> geneExpr[i, ], <span class="at">Tissue =</span> info<span class="sc">$</span>Tissue)</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># Figure 2a</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co"># plot expression stratified by Tissue</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="fu">plotStratify</span>(Expression <span class="sc">~</span> Tissue, GE, <span class="at">main =</span> <span class="fu">rownames</span>(geneExpr)[i])</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/plotStratify-1.png" width="384" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># get gene with the highest variation across Individuals</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co"># create data.frame with expression of gene i and Tissue</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#       type for each sample</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="fu">which.max</span>(varPart<span class="sc">$</span>Individual)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>GE <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>  <span class="at">Expression =</span> geneExpr[i, ],</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>  <span class="at">Individual =</span> info<span class="sc">$</span>Individual</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>)</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co"># Figure 2b</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co"># plot expression stratified by Tissue</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>label <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;Individual:&quot;</span>, <span class="fu">format</span>(varPart<span class="sc">$</span>Individual[i] <span class="sc">*</span> <span class="dv">100</span>,</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>  <span class="at">digits =</span> <span class="dv">3</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>), <span class="st">&quot;%&quot;</span>)</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>main <span class="ot">&lt;-</span> <span class="fu">rownames</span>(geneExpr)[i]</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="fu">plotStratify</span>(Expression <span class="sc">~</span> Individual, GE,</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  <span class="at">colorBy =</span> <span class="cn">NULL</span>,</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  <span class="at">text =</span> label, <span class="at">main =</span> main</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/plotStratify-2.png" width="384" style="display: block; margin: auto;" /></p>
<p>For gene141, variation across tissues explains 52.9% of variance in
gene expression. For gene43, variation across Individuals explains 91.4%
of variance in gene expression.</p>
</div>
<div id="intuition-about-the-backend" class="section level3">
<h3>Intuition about the backend</h3>
<p>At the heart of <code>variancePartition</code>, a regression model is
fit for each gene separately and summary statistics are extracted and
reported to the user for visualization and downstream analysis. For a
single model fit, <code>calcVarPart()</code> computes the fraction of
variance explained by each variable. <code>calcVarPart()</code> is
defined by this package, and computes these statistics from either a
fixed effects model fit with <code>lm()</code> or a linear mixed model
fit with <code>lme4::lmer()</code>.
<code>fitExtractVarPartModel()</code> loops over each gene, fits the
regression model and returns the variance fractions reported by
<code>calcVarPart()</code>.</p>
<p>Fitting the regression model and extracting variance statistics can
also be done directly:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;lme4&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># fit regression model for the first gene</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>form_test <span class="ot">&lt;-</span> geneExpr[<span class="dv">1</span>, ] <span class="sc">~</span> Age <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmer</span>(form_test, info, <span class="at">REML =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co"># extract variance statistics</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="fu">calcVarPart</span>(fit)</span></code></pre></div>
<pre><code>##   Individual       Tissue          Age    Residuals 
## 8.903140e-01 2.468013e-02 4.354738e-05 8.496235e-02</code></pre>
</div>
</div>
<div id="interpretation" class="section level2">
<h2>Interpretation</h2>
<p><code>variancePartition</code> fits a linear (mixed) model that
jointly considers the contribution of all specified variables on the
expression of each gene. It uses a multiple regression model so that the
effect of each variable is assessed while jointly accounting for all
others. <a
href="https://www.r-bloggers.com/2011/03/anova-%E2%80%93-type-iiiiii-ss-explained/">Standard
ANOVA</a> implemented in R involves refitting the model while dropping
terms, but is aimed at hypothesis testing. <code>calcVarPart()</code> is
aimed at estimating variance fractions. It uses a single fit of the
linear (mixed) model and evaluates the sum of squares of each term and
the sum of squares of the total model fit. However, we note that like
any multiple regression model, high correlation bewtween fixed or random
effect variables (see <a
href="#assess-correlation-between-all-pairs-of-variables">Assess
correlation between all pairs of variables</a>) can produce unstable
estimates and it can be challanging to identify which variable is
responsible for the expression variation.</p>
<p>The results of <code>variancePartition</code> give insight into the
expression data at multiple levels. Moreover, a single statistic often
has multiple equivalent interpretations while only one is relevant to
the biological question. Analysis of the example data in Figure 1 gives
some strong interpretations.</p>
<p>Considering the median across all genes,</p>
<ol style="list-style-type: decimal">
<li>variation across individuals explains a median of 81.5% of the
variation in expression, after correcting for tissue, batch and age</li>
<li>variation across tissues explains a median of 8.2% of the variation
in expression, after correcting for other the variables</li>
<li>variation across batches is negligible after correcting for
variation due to other variables</li>
<li>the effect of age is negligible after correcting for other
variables</li>
<li>correcting for individual, tissue, batch and age leaves a median of
9.3% of the total variance in expression.</li>
</ol>
<p>These statistics also have a natural interpretation in terms of the
intra-class correlation (ICC), the correlation between observations made
from samples in the same group.</p>
<p>Considering the median across across all genes and all
experiments,</p>
<ol style="list-style-type: decimal">
<li>the ICC for <u>individual</u> is 81.5%.</li>
<li>the ICC for <u>tissue</u> is 8.2%.</li>
<li>two randomly selected gene measurements from same <u>individual</u>,
but regardless of <u>tissue</u>, <u>batch</u> or <u>age</u>, have a
correlation of 81.5%.</li>
<li>two randomly selected gene measurements from same <u>tissue</u>, but
regardless of <u>individual</u>, <u>batch</u> or <u>age</u>, have a
correlation of 8.2%.</li>
<li>two randomly selected gene measurements from the same
<u>individual</u> <em>and</em> same <u>tissue</u>, but regardless of
<u>batch</u> and <u>age</u>, have an correlation of 81.5% + 8.2% =
89.7%.</li>
</ol>
<p>Note that that the ICC here is interpreted as the ICC after
correcting for all other variables in the model.</p>
<p>These conclusions are based on the genome-wide median across all
genes, but the same type of statements can be made at the gene-level.
Moreover, care must be taken in the interpretation of nested variables.
For example, <code>Age</code> is nested within <code>Individual</code>
since the multiple samples from each individual are taken at the same
age. Thus the effect of <code>Age</code> removes some variation from
being explained by <code>Individual</code>. This often arises when
considering variation across individuals and across sexes: any cross-sex
variation is a component of the cross-individual variation. So the total
variation across individuals is the sum of the fraction of variance
explained by <code>Sex</code> and <code>Individual</code>. This
nesting/summing of effects is common for variables that are properties
of the individual rather than the sample. For example, sex and ethnicity
are always properties of the individual. Variables like age and disease
state can be properties of the individual, but could also vary in
time-course or longitudinal experiments. The the interpretation depends
on the experimental design.</p>
<p>The real power of <code>variancePartition</code> is to identify
specific genes that follow or deviate from the genome-wide trend. The
gene-level statistics can be used to identify a subset of genes that are
enriched for specific biological functions. For example, we can ask if
the 500 genes with the highest variation in expression across tissues
(i.e. the long tail for tissue in Figure 1a) are enriched for genes
known to have high tissue-specificity.</p>
<div id="should-a-variable-be-modeled-as-fixed-or-random-effect"
class="section level3">
<h3>Should a variable be modeled as fixed or random effect?</h3>
<p>Categorical variables should (almost) always be modeled as a random
effect. The difference between modeling a categorical variable as a
fixed versus random effect is minimal when the sample size is large
compared to the number of categories (i.e. levels). So variables like
disease status, sex or time point will not be sensitive to modeling as a
fixed versus random effect. However, variables with many categories like
<code>Individual</code> <em>must</em> be modeled as a random effect in
order to obtain statistically valid results. So to be on the safe side,
categorical variable should be modeled as a random effect.</p>
<p>% R and <code>variancePartition</code> handle catagorical variables
stored as a very naturally. If categorical variables are stored as an or
, they must be converted to a before being used with
<code>variancePartition</code></p>
<p><code>variancePartition</code> fits two types of models:</p>
<ol style="list-style-type: decimal">
<li><p>linear mixed model where <u>all</u> categorical variables are
modeled as random effects and all continuous variables are fixed
effects. The function <code>lme4::lmer()</code> is used to fit this
model.</p></li>
<li><p>fixed effected model, where all variables are modeled as fixed
effects. The function <code>lm()</code> is used to fit this
model.</p></li>
</ol>
</div>
<div id="which-variables-should-be-included" class="section level3">
<h3>Which variables should be included?</h3>
<p>In my experience, it is useful to include all variables in the first
analysis and then drop variables that have minimal effect. However, like
all multiple regression methods, <code>variancePartition</code> will
divide the contribution over multiple variables that are strongly
correlated. So, for example, including both sex and height in the model
will show sex having a smaller contribution to variation gene expression
than if height were omitted, since there variables are strongly
correlated. This is a simple example, but should give some intuition
about a common issue that arises in analyses with
<code>variancePartition</code>.</p>
<p><code>variancePartition</code> can naturally assess the contribution
of both individual and sex in a dataset. As expected, genes for which
sex explains a large fraction of variation are located on chrX and chrY.
If the goal is to interpret the impact of sex, then there is no issue.
But recall the issue with correlated variables and note that individual
is correlated with sex, because each individual is only one sex
regardless of how many samples are taken from a individual. It follows
that including sex in the model reduces the <u>apparent</u> contribution
of individual to gene expression. In other words, the ICC for individual
will be different if sex is included in the model.</p>
<p>In general, including variables in the model that do not vary within
individual will reduce the apparent contribution of individual as
estimated by <code>variancePartition</code>. For example, sex and
ethnicity never vary between multiple samples from the same individual
and will always reduce the apparent contribution of individual. However,
disease state and age may or may not vary depending on the study
design.</p>
<p>In biological datasets technical variability (i.e. batch effects) can
often reduce the apparent biological signal. In RNA-seq analysis, it is
common for the the impact of this technical variability to be removed
before downstream analysis. Instead of including these batch variable in
the <code>variancePartition</code> analysis, it is simple to complete
the expression residuals with the batch effects removed and then feeds
these residuals to <code>variancePartition</code>. This will increase
the fraction of variation explained by biological variables since
technical variability is reduced.</p>
<div id="assess-correlation-between-all-pairs-of-variables"
class="section level4">
<h4>Assess correlation between all pairs of variables</h4>
<p>Evaluating the correlation between variables in a important part in
interpreting variancePartition results. When comparing two continuous
variables, Pearson correlation is widely used. But variancePartition
includes categorical variables in the model as well. In order to
accommodate the correlation between a continuous and a categorical
variable, or two categorical variables we used canonical correlation
analysis.</p>
<p>Canonical Correlation Analysis (CCA) is similar to correlation
between two vectors, except that CCA can accommodate matricies as well.
For a pair of variables, <code>canCorPairs()</code> assesses the degree
to which they co-vary and contain the same information. Variables in the
formula can be a continuous variable or a discrete variable expanded to
a matrix (which is done in the backend of a regression model). For a
pair of variables, <code>canCorPairs()</code> uses CCA to compute the
correlation between these variables and returns the pairwise correlation
matrix.</p>
<p>Statistically, let <code>rho</code> be the array of correlation
values returned by the standard R function cancor to compute CCA.
<code>canCorPairs()</code> returns <code>rho / sum(rho)</code> which is
the fraction of the maximum possible correlation. Note that CCA returns
correlations values between 0 and 1</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> Individual <span class="sc">+</span> Tissue <span class="sc">+</span> Batch <span class="sc">+</span> Age <span class="sc">+</span> Height</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co"># Compute Canonical Correlation Analysis (CCA)</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co"># between all pairs of variables</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co"># returns absolute correlation value</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>C <span class="ot">&lt;-</span> <span class="fu">canCorPairs</span>(form, info)</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co"># Plot correlation matrix</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="co"># between all pairs of variables</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="fu">plotCorrMatrix</span>(C)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/canCorPairs-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="advanced-analysis" class="section level2">
<h2>Advanced analysis</h2>
<div id="extracting-additional-information-from-model-fits"
class="section level3">
<h3>Extracting additional information from model fits</h3>
<p>Advanced users may want to perform the model fit and extract results
in separate steps in order to examine the fit of the model for each
gene. Thus the work of can be divided into two steps: 1) fit the
regression model, and 2) extracting variance statistics.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> Age <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co"># Fit model</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">fitVarPartModel</span>(geneExpr, form, info)</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co"># Extract results</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">extractVarPart</span>(results)</span></code></pre></div>
<p>Note that storing the model fits can use a lot of memory (~10Gb with
20K genes and 1000 experiments). I do not recommend unless you have a
specific need for storing the entire model fit.</p>
<p>Instead, <code>fitVarPartModel()</code> can extract any desired
information using any function that accepts the model fit from
<code>lm()</code> or <code>lmer()</code>. The results are stored in a
<code>list</code> and can be used for downstream analysis.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># Fit model and run summary() function on each model fit</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>vpSummaries <span class="ot">&lt;-</span> <span class="fu">fitVarPartModel</span>(geneExpr, form, info, <span class="at">fxn =</span> summary)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Show results of summary() for the first gene</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>vpSummaries[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## Linear mixed model fit by maximum likelihood  [&#39;lmerMod&#39;]
## Formula: y.local ~ Age + (1 | Individual) + (1 | Tissue) + (1 | Batch)
##    Data: data
## Weights: data$w.local
## Control: control
## 
##      AIC      BIC   logLik deviance df.resid 
##    397.2    412.8   -192.6    385.2       94 
## 
## Scaled residuals: 
##      Min       1Q   Median       3Q      Max 
## -2.05797 -0.58651  0.01466  0.66030  1.97077 
## 
## Random effects:
##  Groups     Name        Variance Std.Dev.
##  Individual (Intercept) 10.82275 3.28979 
##  Batch      (Intercept)  0.00192 0.04382 
##  Tissue     (Intercept)  0.30010 0.54781 
##  Residual                1.02997 1.01488 
## Number of obs: 100, groups:  Individual, 25; Batch, 4; Tissue, 3
## 
## Fixed effects:
##               Estimate Std. Error t value
## (Intercept) -10.602426   1.094925  -9.683
## Age           0.003183   0.016103   0.198
## 
## Correlation of Fixed Effects:
##     (Intr)
## Age -0.739</code></pre>
</div>
<div id="removing-batch-effects-before-fitting-model"
class="section level3">
<h3>Removing batch effects before fitting model</h3>
<p>Gene expression studies often have substantial batch effects, and
<code>variancePartition</code> can be used to understand the magnitude
of the effects. However, we often want to focus on biological variables
(i.e. individual, tissue, disease, sex) after removing the effect of
technical variables. Depending on the size of the batch effect, I have
found it useful to correct for the batch effect first and then perform a
<code>variancePartition</code> analysis afterward. Subtracting this
batch effect can reduce the total variation in the data, so that the
contribution of other variables become clearer.</p>
<p>Standard analysis:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch) <span class="sc">+</span> Age</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(geneExpr, form, info)</span></code></pre></div>
<p>Analysis on residuals:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;limma&quot;</span>)</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="co"># subtract out effect of Batch</span></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(geneExpr, <span class="fu">model.matrix</span>(<span class="sc">~</span>Batch, info))</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">residuals</span>(fit, geneExpr)</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="co"># fit model on residuals</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> Age</span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a>varPartResid <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(res, form, info)</span></code></pre></div>
<p>Remove batch effect with linear mixed model</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># subtract out effect of Batch with linear mixed model</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>modelFit <span class="ot">&lt;-</span> <span class="fu">fitVarPartModel</span>(geneExpr, <span class="sc">~</span> (<span class="dv">1</span> <span class="sc">|</span> Batch), info)</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">residuals</span>(modelFit)</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="co"># fit model on residuals</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> Age</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a>varPartResid <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(res, form, info)</span></code></pre></div>
<p>If the two-step process requires too much memory, the residuals can
be computed more efficiently. Here, run the function inside the call to
<code>fitVarPartModel()</code> to avoid storing the large intermediate
results.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># extract residuals directly without storing intermediate results</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>residList <span class="ot">&lt;-</span> <span class="fu">fitVarPartModel</span>(geneExpr, <span class="sc">~</span> (<span class="dv">1</span> <span class="sc">|</span> Batch), info,</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>  <span class="at">fxn =</span> residuals</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>)</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a><span class="co"># convert list to matrix</span></span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>residMatrix <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, residList)</span></code></pre></div>
</div>
<div id="variation-within-multiple-subsets-of-the-data"
class="section level3">
<h3>Variation within multiple subsets of the data</h3>
<p>So far, we have focused on interpreting one variable at a time. But
the linear mixed model behind <code>variancePartition</code> is a very
powerful framework for analyzing variation at multiple levels. We can
easily extend the previous analysis of the contribution of individual
and tissue on variation in gene expression to examine the contribution
of individual <em>within</em> each tissue. This analysis is as easy as
specifying a new formula and rerunning <code>variancePartition</code>.
Note that is analysis will only work when there are replicates for at
least some individuals within each tissue in order to assess
cross-individual variance with in a tissue.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># specify formula to model within/between individual variance</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="co"># separately for each tissue</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="co"># Note that including +0 ensures each tissue is modeled explicitly</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a><span class="co"># Otherwise, the first tissue would be used as baseline</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (Tissue <span class="sc">+</span> <span class="dv">0</span> <span class="sc">|</span> Individual) <span class="sc">+</span> Age <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch)</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a><span class="co"># fit model and extract variance percents</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(geneExpr, form, info, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a><span class="co"># violin plot</span></span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a><span class="fu">plotVarPart</span>(<span class="fu">sortCols</span>(varPart), <span class="at">label.angle =</span> <span class="dv">60</span>)</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/withinTissue-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>This analysis corresponds to a varying coefficient model, where the
correlation between individuals varies for each tissue ]. Since the
variation across individuals is modeled within each tissue, the total
variation explained does not sum to 1 as it does for standard
application of <code>variancePartition</code>. So interpretation as
intra-class does not strictly apply and use of
<code>plotPercentBars()</code> is no longer applicable. Yet the
variables in the study design are still ranked in terms of their
genome-wide contribution to expression variation, and results can still
be analyzed at the gene level. See <a
href="#variation-with-multiple-subsets-of-the-data">Variation with
multiple subsets of the data</a> for statistical details.</p>
</div>
<div id="detecting-problems-caused-by-collinearity-of-variables"
class="section level3">
<h3>Detecting problems caused by collinearity of variables</h3>
<p>Including variables that are highly correlated can produce misleading
results and overestimate the contribution of variables modeled as fixed
effects. This is usually not an issue, but can arise when statistically
redundant variables are included in the model. In this case, the model
is <code>"degenerate"</code> or <code>"computationally singular"</code>
and parameter estimates from this model are not meaningful. Dropping one
or more of the covariates will fix this problem.</p>
<p>A check of collinearity is built into <code>fitVarPartModel()</code>
and <code>fitExtractVarPartModel()</code>, so the user will be warned if
this is an issue.</p>
<p>Alternatively, the user can use the <code>colinearityScore()</code>
function to evaluate whether this is an issue for a single model
fit:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> Age <span class="sc">+</span> Height</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a><span class="co"># fit model</span></span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">fitVarPartModel</span>(geneExpr[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, ], form, info)</span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co"># evaluate the collinearity score on the first model fit</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="co"># this reports the correlation matrix between coefficient estimates</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="co"># for fixed effects</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a><span class="co"># the collinearity score is the maximum absolute correlation value</span></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a><span class="co"># If the collinearity score &gt; .99 then the variance partition</span></span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a><span class="co"># estimates may be problematic</span></span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a><span class="co"># In that case, a least one variable should be omitted</span></span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a><span class="fu">colinearityScore</span>(res[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>## [1] 0.7774082
## attr(,&quot;vcor&quot;)
##             (Intercept)         Age      Height
## (Intercept)   1.0000000 -0.41907978 -0.77740819
## Age          -0.4190798  1.00000000 -0.05748224
## Height       -0.7774082 -0.05748224  1.00000000</code></pre>
</div>
<div id="including-weights-computed-separately" class="section level3">
<h3>Including weights computed separately</h3>
<p><code>variancePartition</code> automatically used precision weights
computed by <code>limma::voom()</code>, but the user can also specify
custom weights using the <code>weightsMatrix</code> argument.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> Age <span class="sc">+</span> Height</span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a><span class="co"># Specify custom weights</span></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a><span class="co"># In this example the weights are simulated from a</span></span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a><span class="co"># uniform distribution and are not meaningful.</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>weights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">runif</span>(<span class="fu">length</span>(geneExpr)), <span class="at">nrow =</span> <span class="fu">nrow</span>(geneExpr))</span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a><span class="co"># Specify custom weights</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(geneExpr[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, ], form, info,</span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a>  <span class="at">weightsMatrix =</span> weights[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, ]</span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a>)</span></code></pre></div>
<p>In addition, setting the <code>useWeights=FALSE</code> will suppress
usage of the weights in all cases, i.e. when the weights are specified
manually or implicitly with the results of
<code>limma::voom()</code>.</p>
</div>
<div id="including-interaction-terms" class="section level3">
<h3>Including interaction terms</h3>
<p>Typical analysis assumes that the effect of each variable on gene
expression does not depend on other variables in the model. Sometimes
this assumption is too strict, and we want to model an interaction
effect whereby the effect of <code>Batch</code> depends on
<code>Tissue</code>. This can be done easly by specifying an interaction
term, <code>(1|Batch:Tissue)</code>. Since <code>Batch</code> has 4
categories and <code>Tissue</code> has 3, this interaction term
implicity models a new <code>3*4 = 12</code> category variable in the
analysis. This new interaction term will absorb some of the variance
from the <code>Batch</code> and <code>Tissue</code> term, so an
interaction model should always include the two constituent
variables.</p>
<p>Here we fit an interaction model, but we observe that interaction
between <code>Batch</code> and <code>Tissue</code> does not explain much
expression variation.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> Age <span class="sc">+</span> Height <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch) <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>  (<span class="dv">1</span> <span class="sc">|</span> Batch<span class="sc">:</span>Tissue)</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a><span class="co"># fit model</span></span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>vpInteraction <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(geneExpr, form, info)</span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a><span class="fu">plotVarPart</span>(<span class="fu">sortCols</span>(vpInteraction))</span></code></pre></div>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/vpInteraction-1.png" width="384" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="application-to-expression-data" class="section level1">
<h1>Application to expression data</h1>
<p><code>variancePartition</code> works with gene expression data that
has already been processed and normalized as for differential expression
analysis.</p>
<div id="gene-level-counts" class="section level2">
<h2>Gene-level counts</h2>
<p><code>featureCounts</code> <span class="citation">(Liao, Smyth, and
Shi 2014)</span> and <code>HTSeq</code> <span class="citation">(Anders,
Pyl, and Huber 2015)</span> report the number of reads mapping to each
gene (or exon). These results are easily read into R.
<code>limma::voom()</code> and <code>DESeq2</code> are widely used for
differential expression analysis of gene- and exon-level counts and can
be used to process data before analysis with
<code>variancePartition</code>. This section addresses processing and
normalization of gene-level counts, but the analysis is the same for
exon-level counts.</p>
<div id="limmavoom" class="section level3">
<h3><code>limma::voom()</code></h3>
<p>Read RNA-seq counts into R, normalize for library size within and
between experiments with TMM <span class="citation">(Robinson and
Oshlack 2010)</span>, estimate precision weights with
<code>limma::voom()</code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;limma&quot;</span>)</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;edgeR&quot;</span>)</span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a><span class="co"># identify genes that pass expression cutoff</span></span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>isexpr <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">cpm</span>(geneCounts) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">ncol</span>(geneCounts)</span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a></span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a><span class="co"># create data structure with only expressed genes</span></span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a>gExpr <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts =</span> geneCounts[isexpr, ])</span>
<span id="cb29-9"><a href="#cb29-9" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" tabindex="-1"></a><span class="co"># Perform TMM normalization</span></span>
<span id="cb29-11"><a href="#cb29-11" tabindex="-1"></a>gExpr <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(gExpr)</span>
<span id="cb29-12"><a href="#cb29-12" tabindex="-1"></a></span>
<span id="cb29-13"><a href="#cb29-13" tabindex="-1"></a><span class="co"># Specify variables to be included in the voom() estimates of</span></span>
<span id="cb29-14"><a href="#cb29-14" tabindex="-1"></a><span class="co"># uncertainty.</span></span>
<span id="cb29-15"><a href="#cb29-15" tabindex="-1"></a><span class="co"># Recommend including variables with a small number of categories</span></span>
<span id="cb29-16"><a href="#cb29-16" tabindex="-1"></a><span class="co"># that explain a substantial amount of variation</span></span>
<span id="cb29-17"><a href="#cb29-17" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>Batch, info)</span>
<span id="cb29-18"><a href="#cb29-18" tabindex="-1"></a></span>
<span id="cb29-19"><a href="#cb29-19" tabindex="-1"></a><span class="co"># Estimate precision weights for each gene and sample</span></span>
<span id="cb29-20"><a href="#cb29-20" tabindex="-1"></a><span class="co"># This models uncertainty in expression measurements</span></span>
<span id="cb29-21"><a href="#cb29-21" tabindex="-1"></a>vobjGenes <span class="ot">&lt;-</span> <span class="fu">voom</span>(gExpr, design)</span>
<span id="cb29-22"><a href="#cb29-22" tabindex="-1"></a></span>
<span id="cb29-23"><a href="#cb29-23" tabindex="-1"></a><span class="co"># Define formula</span></span>
<span id="cb29-24"><a href="#cb29-24" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch) <span class="sc">+</span> Age</span>
<span id="cb29-25"><a href="#cb29-25" tabindex="-1"></a></span>
<span id="cb29-26"><a href="#cb29-26" tabindex="-1"></a><span class="co"># variancePartition seamlessly deals with the result of voom()</span></span>
<span id="cb29-27"><a href="#cb29-27" tabindex="-1"></a><span class="co"># by default, it seamlessly models the precision weights</span></span>
<span id="cb29-28"><a href="#cb29-28" tabindex="-1"></a><span class="co"># This can be turned off with useWeights=FALSE</span></span>
<span id="cb29-29"><a href="#cb29-29" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(vobjGenes, form, info)</span></code></pre></div>
</div>
<div id="deseq2" class="section level3">
<h3><code>DESeq2</code></h3>
<p>Process and normalize the gene-level counts before running
<code>variancePartition</code> analysis.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;DESeq2&quot;</span>)</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a><span class="co"># create DESeq2 object from gene-level counts and metadata</span></span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb30-5"><a href="#cb30-5" tabindex="-1"></a>  <span class="at">countData =</span> geneCounts,</span>
<span id="cb30-6"><a href="#cb30-6" tabindex="-1"></a>  <span class="at">colData =</span> info,</span>
<span id="cb30-7"><a href="#cb30-7" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span><span class="dv">1</span></span>
<span id="cb30-8"><a href="#cb30-8" tabindex="-1"></a>)</span>
<span id="cb30-9"><a href="#cb30-9" tabindex="-1"></a></span>
<span id="cb30-10"><a href="#cb30-10" tabindex="-1"></a><span class="co"># Estimate library size correction scaling factors</span></span>
<span id="cb30-11"><a href="#cb30-11" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb30-12"><a href="#cb30-12" tabindex="-1"></a></span>
<span id="cb30-13"><a href="#cb30-13" tabindex="-1"></a><span class="co"># identify genes that pass expression cutoff</span></span>
<span id="cb30-14"><a href="#cb30-14" tabindex="-1"></a>isexpr <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">fpm</span>(dds) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">ncol</span>(dds)</span>
<span id="cb30-15"><a href="#cb30-15" tabindex="-1"></a></span>
<span id="cb30-16"><a href="#cb30-16" tabindex="-1"></a><span class="co"># compute log2 Fragments Per Million</span></span>
<span id="cb30-17"><a href="#cb30-17" tabindex="-1"></a><span class="co"># Alternatively, fpkm(), vst() or rlog() could be used</span></span>
<span id="cb30-18"><a href="#cb30-18" tabindex="-1"></a>quantLog <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fu">fpm</span>(dds)[isexpr, ] <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb30-19"><a href="#cb30-19" tabindex="-1"></a></span>
<span id="cb30-20"><a href="#cb30-20" tabindex="-1"></a><span class="co"># Define formula</span></span>
<span id="cb30-21"><a href="#cb30-21" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Individual) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Tissue) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Batch) <span class="sc">+</span> Age</span>
<span id="cb30-22"><a href="#cb30-22" tabindex="-1"></a></span>
<span id="cb30-23"><a href="#cb30-23" tabindex="-1"></a><span class="co"># Run variancePartition analysis</span></span>
<span id="cb30-24"><a href="#cb30-24" tabindex="-1"></a>varPart <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(quantLog, form, info)</span></code></pre></div>
<p>Note that <code>DESeq2</code> does not compute precision weights like
<code>limma::voom()</code>, so they are not used in this version of the
analysis.</p>
</div>
<div id="isoform-quantification" class="section level3">
<h3>Isoform quantification</h3>
<p>Other software performs isoform-level quantification using reads that
map to multiple transcripts. These include <code>kallisto</code> <span
class="citation">(Bray et al. 2016)</span>, <code>sailfish</code> <span
class="citation">(Patro, Mount, and Kingsford 2014)</span>,
<code>salmon</code> <span class="citation">Patro, Duggal, and Kingsford
(2015)</span>, <code>RSEM</code> <span class="citation">(Li and Dewey
2011)</span>and <code>cufflinks</code> <span class="citation">(Trapnell
et al. 2010.)</span></p>
<div id="tximport" class="section level4">
<h4><code>tximport</code></h4>
<p>Quantifications from <code>kallisto</code>, <code>salmon</code>,
<code>sailfish</code> and <code>RSEM</code> can be read into R and
processed with the Bioconductor package <code>tximport</code>. The gene-
or transcript-level quantifications can be used directly in
<code>variancePartition</code>.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;tximportData&quot;</span>)</span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;tximport&quot;</span>)</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;readr&quot;</span>)</span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a><span class="co"># Get data from folder where tximportData is installed</span></span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a>dir <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="at">package =</span> <span class="st">&quot;tximportData&quot;</span>)</span>
<span id="cb31-7"><a href="#cb31-7" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">file.path</span>(dir, <span class="st">&quot;samples.txt&quot;</span>), <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-8"><a href="#cb31-8" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dir, <span class="st">&quot;kallisto&quot;</span>, samples<span class="sc">$</span>run, <span class="st">&quot;abundance.tsv&quot;</span>)</span>
<span id="cb31-9"><a href="#cb31-9" tabindex="-1"></a><span class="fu">names</span>(files) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;sample&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb31-10"><a href="#cb31-10" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" tabindex="-1"></a>tx2gene <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">file.path</span>(dir, <span class="st">&quot;tx2gene.csv&quot;</span>))</span>
<span id="cb31-12"><a href="#cb31-12" tabindex="-1"></a></span>
<span id="cb31-13"><a href="#cb31-13" tabindex="-1"></a><span class="co"># reads results from kallisto</span></span>
<span id="cb31-14"><a href="#cb31-14" tabindex="-1"></a>txi <span class="ot">&lt;-</span> <span class="fu">tximport</span>(files,</span>
<span id="cb31-15"><a href="#cb31-15" tabindex="-1"></a>  <span class="at">type =</span> <span class="st">&quot;kallisto&quot;</span>, <span class="at">tx2gene =</span> tx2gene,</span>
<span id="cb31-16"><a href="#cb31-16" tabindex="-1"></a>  <span class="at">countsFromAbundance =</span> <span class="st">&quot;lengthScaledTPM&quot;</span></span>
<span id="cb31-17"><a href="#cb31-17" tabindex="-1"></a>)</span>
<span id="cb31-18"><a href="#cb31-18" tabindex="-1"></a></span>
<span id="cb31-19"><a href="#cb31-19" tabindex="-1"></a><span class="co"># define metadata (usually read from external source)</span></span>
<span id="cb31-20"><a href="#cb31-20" tabindex="-1"></a>info_tximport <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb31-21"><a href="#cb31-21" tabindex="-1"></a>  <span class="at">Sample =</span> <span class="fu">sprintf</span>(<span class="st">&quot;sample%d&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>),</span>
<span id="cb31-22"><a href="#cb31-22" tabindex="-1"></a>  <span class="at">Disease =</span> <span class="fu">c</span>(<span class="st">&quot;case&quot;</span>, <span class="st">&quot;control&quot;</span>)[<span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="fu">rep</span>(<span class="dv">2</span>, <span class="dv">3</span>))]</span>
<span id="cb31-23"><a href="#cb31-23" tabindex="-1"></a>)</span>
<span id="cb31-24"><a href="#cb31-24" tabindex="-1"></a></span>
<span id="cb31-25"><a href="#cb31-25" tabindex="-1"></a><span class="co"># Extract counts from kallisto</span></span>
<span id="cb31-26"><a href="#cb31-26" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(txi<span class="sc">$</span>counts)</span>
<span id="cb31-27"><a href="#cb31-27" tabindex="-1"></a></span>
<span id="cb31-28"><a href="#cb31-28" tabindex="-1"></a><span class="co"># compute library size normalization</span></span>
<span id="cb31-29"><a href="#cb31-29" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(y)</span>
<span id="cb31-30"><a href="#cb31-30" tabindex="-1"></a></span>
<span id="cb31-31"><a href="#cb31-31" tabindex="-1"></a><span class="co"># apply voom to estimate precision weights</span></span>
<span id="cb31-32"><a href="#cb31-32" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>Disease, <span class="at">data =</span> info_tximport)</span>
<span id="cb31-33"><a href="#cb31-33" tabindex="-1"></a>vobj <span class="ot">&lt;-</span> <span class="fu">voom</span>(y, design)</span>
<span id="cb31-34"><a href="#cb31-34" tabindex="-1"></a></span>
<span id="cb31-35"><a href="#cb31-35" tabindex="-1"></a><span class="co"># define formula</span></span>
<span id="cb31-36"><a href="#cb31-36" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Disease)</span>
<span id="cb31-37"><a href="#cb31-37" tabindex="-1"></a></span>
<span id="cb31-38"><a href="#cb31-38" tabindex="-1"></a><span class="co"># Run variancePartition analysis (on only 10 genes)</span></span>
<span id="cb31-39"><a href="#cb31-39" tabindex="-1"></a>varPart_tx <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(</span>
<span id="cb31-40"><a href="#cb31-40" tabindex="-1"></a>  vobj[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ], form,</span>
<span id="cb31-41"><a href="#cb31-41" tabindex="-1"></a>  info_tximport</span>
<span id="cb31-42"><a href="#cb31-42" tabindex="-1"></a>)</span></code></pre></div>
<p>Code to process results from <code>sailfish</code>,
<code>salmon</code>, <code>RSEM</code> is very similar.</p>
<p>See <a
href="http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html">tutorial</a>
for more details.</p>
</div>
</div>
<div id="ballgown" class="section level3">
<h3><code>ballgown</code></h3>
<p>Quantifications from Cufflinks/Tablemaker and RSEM can be processed
and read into R with the Bioconductor package <code>ballgown</code>.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;ballgown&quot;</span>)</span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a><span class="co"># Get data from folder where ballgown is installed</span></span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a>data_directory <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="at">package =</span> <span class="st">&quot;ballgown&quot;</span>)</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" tabindex="-1"></a><span class="co"># Load results of Cufflinks/Tablemaker</span></span>
<span id="cb32-7"><a href="#cb32-7" tabindex="-1"></a>bg <span class="ot">&lt;-</span> <span class="fu">ballgown</span>(</span>
<span id="cb32-8"><a href="#cb32-8" tabindex="-1"></a>  <span class="at">dataDir =</span> data_directory, <span class="at">samplePattern =</span> <span class="st">&quot;sample&quot;</span>,</span>
<span id="cb32-9"><a href="#cb32-9" tabindex="-1"></a>  <span class="at">meas =</span> <span class="st">&quot;all&quot;</span></span>
<span id="cb32-10"><a href="#cb32-10" tabindex="-1"></a>)</span>
<span id="cb32-11"><a href="#cb32-11" tabindex="-1"></a></span>
<span id="cb32-12"><a href="#cb32-12" tabindex="-1"></a><span class="co"># extract gene-level FPKM quantification</span></span>
<span id="cb32-13"><a href="#cb32-13" tabindex="-1"></a><span class="co"># Expression can be convert to log2-scale if desired</span></span>
<span id="cb32-14"><a href="#cb32-14" tabindex="-1"></a>gene_expression <span class="ot">&lt;-</span> <span class="fu">gexpr</span>(bg)</span>
<span id="cb32-15"><a href="#cb32-15" tabindex="-1"></a></span>
<span id="cb32-16"><a href="#cb32-16" tabindex="-1"></a><span class="co"># extract transcript-level FPKM quantification</span></span>
<span id="cb32-17"><a href="#cb32-17" tabindex="-1"></a><span class="co"># Expression can be convert to log2-scale if desired</span></span>
<span id="cb32-18"><a href="#cb32-18" tabindex="-1"></a>transcript_fpkm <span class="ot">&lt;-</span> <span class="fu">texpr</span>(bg, <span class="st">&quot;FPKM&quot;</span>)</span>
<span id="cb32-19"><a href="#cb32-19" tabindex="-1"></a></span>
<span id="cb32-20"><a href="#cb32-20" tabindex="-1"></a><span class="co"># define metadata (usually read from external source)</span></span>
<span id="cb32-21"><a href="#cb32-21" tabindex="-1"></a>info_ballgown <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb32-22"><a href="#cb32-22" tabindex="-1"></a>  <span class="at">Sample =</span> <span class="fu">sprintf</span>(<span class="st">&quot;sample%02d&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>),</span>
<span id="cb32-23"><a href="#cb32-23" tabindex="-1"></a>  <span class="at">Batch =</span> <span class="fu">rep</span>(letters[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="dv">5</span>),</span>
<span id="cb32-24"><a href="#cb32-24" tabindex="-1"></a>  <span class="at">Disease =</span> <span class="fu">c</span>(<span class="st">&quot;case&quot;</span>, <span class="st">&quot;control&quot;</span>)[<span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">10</span>), <span class="fu">rep</span>(<span class="dv">2</span>, <span class="dv">10</span>))]</span>
<span id="cb32-25"><a href="#cb32-25" tabindex="-1"></a>)</span>
<span id="cb32-26"><a href="#cb32-26" tabindex="-1"></a></span>
<span id="cb32-27"><a href="#cb32-27" tabindex="-1"></a><span class="co"># define formula</span></span>
<span id="cb32-28"><a href="#cb32-28" tabindex="-1"></a>form <span class="ot">&lt;-</span> <span class="er">~</span> (<span class="dv">1</span> <span class="sc">|</span> Batch) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> Disease)</span>
<span id="cb32-29"><a href="#cb32-29" tabindex="-1"></a></span>
<span id="cb32-30"><a href="#cb32-30" tabindex="-1"></a><span class="co"># Run variancePartition analysis</span></span>
<span id="cb32-31"><a href="#cb32-31" tabindex="-1"></a><span class="co"># Gene-level analysis</span></span>
<span id="cb32-32"><a href="#cb32-32" tabindex="-1"></a>varPart_gene <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(</span>
<span id="cb32-33"><a href="#cb32-33" tabindex="-1"></a>  gene_expression, form,</span>
<span id="cb32-34"><a href="#cb32-34" tabindex="-1"></a>  info_ballgown</span>
<span id="cb32-35"><a href="#cb32-35" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Warning in filterInputData(exprObj, formula, data, useWeights = useWeights): Sample names of responses (i.e. columns of exprObj) do not match
## sample names of metadata (i.e. rows of data).  Recommend consistent
## names so downstream results are labeled consistently.</code></pre>
<pre><code>## Warning in .fitExtractVarPartModel(exprObj, formula, data, REML = REML, : Model failed for 1 responses.
##   See errors with attr(., &#39;errors&#39;)</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="co"># Transcript-level analysis</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a>varPart_transcript <span class="ot">&lt;-</span> <span class="fu">fitExtractVarPartModel</span>(</span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a>  transcript_fpkm, form,</span>
<span id="cb35-4"><a href="#cb35-4" tabindex="-1"></a>  info_ballgown</span>
<span id="cb35-5"><a href="#cb35-5" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Warning in filterInputData(exprObj, formula, data, useWeights = useWeights): Sample names of responses (i.e. columns of exprObj) do not match
## sample names of metadata (i.e. rows of data).  Recommend consistent
## names so downstream results are labeled consistently.</code></pre>
<pre><code>## Warning in .fitExtractVarPartModel(exprObj, formula, data, REML = REML, : Model failed for 2 responses.
##   See errors with attr(., &#39;errors&#39;)</code></pre>
<p>Note that <code>ballgownrsem</code> can be used for a similar
analysis of <code>RSEM</code> results.</p>
<p>See <a
href="http://bioconductor.org/packages/release/bioc/vignettes/ballgown/inst/doc/ballgown.html">tutorial</a>
for more details.</p>
</div>
</div>
<div id="compare-with-other-methods" class="section level2">
<h2>Compare with other methods</h2>
<p>Characterizing drivers of variation in gene expression data has
typically relied on principal components analysis (PCA) and hierarchical
clustering. Here I apply these methods to two simulated datasets to
demonstrate the additional insight from an analysis with
<code>variancePartition</code>. Each simulated dataset comprises 60
experiments from 10 individuals and 3 tissues with 2 biological
replicates. In the first dataset, tissue is the major driver of
variation in gene expression(Figure @ref(fig:siteDominant)). In the
second dataset, individual is the major driver of variation in gene
expression (Figures @ref(fig:IndivDominant)).</p>
<p>Analysis of simulated data illustrates that PCA identifies the major
driver of variation when tissue is dominant and there are only 3
categories. But the results are less clear when individual is dominant
because there are now 10 categories. Meanwhile, hierarchical clustering
identifies the major driver of variation in both cases, but does not
give insight into the second leading contributor.</p>
<p>Analysis with <code>variancePartition</code> has a number of
advantages over these standard methods:</p>
<ol style="list-style-type: decimal">
<li><p><code>variancePartition</code> provides a natural interpretation
of multiple variables</p>
<ul>
<li>figures from PCA/hierarchical clustering allow easy interpretation
of only one variable</li>
</ul></li>
<li><p><code>variancePartition</code> quantifies the contribution of
each variable</p>
<ul>
<li>PCA/hierarchical clustering give only a visual representation</li>
</ul></li>
<li><p><code>variancePartition</code> interprets contribution of each
variable to each gene individually for downstream analysis</p>
<ul>
<li>PCA/hierarchical clustering produces genome-wide summary and does
not allow gene-level interpretation</li>
</ul></li>
<li><p><code>variancePartition</code> can assess contribution of one
variable (i.e. Individual) separately in subset of the data defined by
another variable (i.e. Tissue)</p></li>
</ol>
<div id="tissue-is-major-source-of-variation" class="section level3">
<h3>Tissue is major source of variation</h3>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/siteDominant-1.png" width="768" style="display: block; margin: auto;" />
<img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/cowplot-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="individual-is-major-source-of-variation"
class="section level3">
<h3>Individual is major source of variation</h3>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/IndivDominant-1.png" width="768" style="display: block; margin: auto;" />
<img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/cowplot2-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="statistical-details" class="section level2">
<h2>Statistical details</h2>
<p>A <code>variancePartition</code> analysis evaluates the linear
(mixed) model
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>y</mi></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>X</mi><mi>j</mi></msub><msub><mi>β</mi><mi>j</mi></msub><mo>+</mo><munder><mo>∑</mo><mi>k</mi></munder><msub><mi>Z</mi><mi>k</mi></msub><msub><mi>α</mi><mi>k</mi></msub><mo>+</mo><mi>ε</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>α</mi><mi>k</mi></msub></mtd><mtd columnalign="center" style="text-align: center"><mo>∼</mo></mtd><mtd columnalign="left" style="text-align: left"><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><msubsup><mi>σ</mi><msub><mi>α</mi><mi>k</mi></msub><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>ε</mi></mtd><mtd columnalign="center" style="text-align: center"><mo>∼</mo></mtd><mtd columnalign="left" style="text-align: left"><mi>𝒩</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
  y &amp;=&amp; \sum_j X_j\beta_j + \sum_k Z_k \alpha_k + \varepsilon \\
  \alpha_k &amp;\sim&amp; \mathcal{N}(0, \sigma^2_{\alpha_k})\\
  \varepsilon &amp;\sim&amp; \mathcal{N}(0, \sigma^2_\varepsilon)
\end{eqnarray}</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
is the expression of a single gene across all samples,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mi>j</mi></msub><annotation encoding="application/x-tex">X_j</annotation></semantics></math>
is the matrix of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>j</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">j^{th}</annotation></semantics></math>
fixed effect with coefficients
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>β</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\beta_j</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Z</mi><mi>k</mi></msub><annotation encoding="application/x-tex">Z_k</annotation></semantics></math>
is the matrix corresponding to the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>k</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">k^{th}</annotation></semantics></math>
random effect with coefficients
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>α</mi><mi>k</mi></msub><annotation encoding="application/x-tex">\alpha_k</annotation></semantics></math>
drawn from a normal distribution with variance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><msub><mi>α</mi><mi>k</mi></msub><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{\alpha_k}</annotation></semantics></math>.
The noise term,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ε</mi><annotation encoding="application/x-tex">\varepsilon</annotation></semantics></math>,
is drawn from a normal distribution with variance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_\varepsilon</annotation></semantics></math>.
Parameters are estimated with maximum likelihood, rather than REML, so
that fixed effect coefficients,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>β</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\beta_j</annotation></semantics></math>,
are explicitly estimated.</p>
<p>I use the term “linear (mixed) model” here since
<code>variancePartition</code> works seamlessly when a fixed effects
model (i.e. linear model) is specified.</p>
<p>Variance terms for the fixed effects are computed using the <em>post
hoc</em> calculation<br />
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>β</mi><mi>j</mi></msub><mn>2</mn></msubsup><mo>=</mo><mtext mathvariant="normal">var</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>X</mi><mi>j</mi></msub><msub><mover><mi>β</mi><mo accent="true">̂</mo></mover><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
 \hat{\sigma}^2_{\beta_j} = \text{var}\left( X_j \hat{\beta}_j\right).
\end{eqnarray}</annotation></semantics></math> For a fixed effects
model, this corresponds to the sum of squares for each component of the
model.</p>
<p>For a standard application of the linear mixed model, where the
effect of each variable is additive, the fraction of variance explained
by the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>j</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">j^{th}</annotation></semantics></math>
fixed effect is
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mfrac><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>β</mi><mi>j</mi></msub><mn>2</mn></msubsup><mrow><munder><mo>∑</mo><mi>j</mi></munder><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>β</mi><mi>j</mi></msub><mn>2</mn></msubsup><mo>+</mo><munder><mo>∑</mo><mi>k</mi></munder><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>α</mi><mi>k</mi></msub><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><mo>,</mo></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
 \frac{\hat{\sigma}^2_{\beta_j}}{\sum_j \hat{\sigma}^2_{\beta_j} + \sum_k \hat{\sigma}^2_{\alpha_k} + \hat{\sigma}^2_\varepsilon},
\end{eqnarray}</annotation></semantics></math> by the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>k</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">k^{th}</annotation></semantics></math>
random effect is
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mfrac><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>α</mi><mi>k</mi></msub><mn>2</mn></msubsup><mrow><munder><mo>∑</mo><mi>j</mi></munder><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>β</mi><mi>j</mi></msub><mn>2</mn></msubsup><mo>+</mo><munder><mo>∑</mo><mi>k</mi></munder><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>α</mi><mi>k</mi></msub><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><mo>,</mo></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
 \frac{\hat{\sigma}^2_{\alpha_k}}{\sum_j \hat{\sigma}^2_{\beta_j} + \sum_k \hat{\sigma}^2_{\alpha_k} + \hat{\sigma}^2_\varepsilon},
 \end{eqnarray}</annotation></semantics></math> and the residual
variance is
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mfrac><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>ε</mi><mn>2</mn></msubsup><mrow><munder><mo>∑</mo><mi>j</mi></munder><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>β</mi><mi>j</mi></msub><mn>2</mn></msubsup><mo>+</mo><munder><mo>∑</mo><mi>k</mi></munder><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><msub><mi>α</mi><mi>k</mi></msub><mn>2</mn></msubsup><mo>+</mo><msubsup><mover><mi>σ</mi><mo accent="true">̂</mo></mover><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><mi>.</mi></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
 \frac{\hat{\sigma}^2_{\varepsilon}}{\sum_j \hat{\sigma}^2_{\beta_j} + \sum_k \hat{\sigma}^2_{\alpha_k} + \hat{\sigma}^2_\varepsilon}.
 \end{eqnarray}</annotation></semantics></math></p>
<div id="implementation-in-r" class="section level3">
<h3>Implementation in R</h3>
<p>An R formula is used to define the terms in the fixed and random
effects, and <code>fitVarPartModel()</code> fits the specified model for
each gene separately. If random effects are specified,
<code>lme4::lmer()</code> is used behind the scenes to fit the model,
while <code>lm()</code> is used if there are only fixed effects.
<code>fitVarPartModel()</code> returns a list of the model fits, and
returns the variance partition statistics for each model in the list.
<code>fitExtractVarPartModel()</code> combines the actions of
<code>fitVarPartModel()</code> and into one function call.
<code>calcVarPart()</code> is called behind the scenes to compute
variance fractions for both fixed and mixed effects models, but the user
can also call this function directly on a model fit with
<code>lm()</code> or <code>lmer()</code>.</p>
</div>
<div id="interpretation-of-percent-variance-explained"
class="section level3">
<h3>Interpretation of percent variance explained</h3>
<p>The percent variance explained can be interpreted as the intra-class
correlation (ICC) when a special case of Equation 1 is used. Consider
the simplest example of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>i</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">i^{th}</annotation></semantics></math>
sample from the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>k</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">k^{th}</annotation></semantics></math>
individual
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi></mrow></msub><mo>=</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi></mrow></msub></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
  y_{i,k} = \mu + Z \alpha_{i,k} + e_{i,k}
\end{eqnarray}</annotation></semantics></math> with only an intercept
term, one random effect corresponding to individual, and an error term.
In this case ICC corresponds to the correlation between two samples from
the same individual. This value is equal to the fraction of variance
explained by individual. For example, consider the correlation between
samples from the same individual:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>I</mi><mi>C</mi><mi>C</mi></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mi>c</mi><mi>o</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>,</mo><msub><mi>y</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mi>c</mi><mi>o</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>,</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mfrac><mrow><mi>c</mi><mi>o</mi><mi>v</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>,</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><msqrt><mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mi>v</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msqrt></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mfrac><mrow><mi>c</mi><mi>o</mi><mi>v</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mi>k</mi></mrow></msub><mo>,</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>2</mn><mo>,</mo><mi>k</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><msubsup><mi>σ</mi><mi>α</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mfrac><msubsup><mi>σ</mi><mi>α</mi><mn>2</mn></msubsup><mrow><msubsup><mi>σ</mi><mi>α</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
  ICC &amp;=&amp; cor( y_{1,k}, y_{2,k}) \\
      &amp;=&amp; cor( \mu + Z \alpha_{1,k}  + e_{1,k}, \mu + Z \alpha_{2,k}  + e_{2,k}) \\
      &amp;=&amp; \frac{cov( \mu + Z \alpha_{1,k}  + e_{1,k}, \mu + Z \alpha_{2,k}  + e_{2,k})}{ \sqrt{ var(\mu + Z \alpha_{1,k} + e_{1,k}) var( \mu + Z \alpha_{2,k} + e_{2,k})}}\\
      &amp;=&amp; \frac{cov(Z \alpha_{1,k}, Z \alpha_{2,k})}{\sigma^2_\alpha + \sigma^2_\varepsilon} \\
      &amp;=&amp; \frac{\sigma^2_\alpha}{\sigma^2_\alpha + \sigma^2_\varepsilon}
\end{eqnarray}</annotation></semantics></math> The correlation between
samples from different individuals is:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mi>c</mi><mi>o</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mrow><mn>1</mn><mo>,</mo><mn>1</mn></mrow></msub><mo>,</mo><msub><mi>y</mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mi>c</mi><mi>o</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mn>1</mn></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>1</mn><mo>,</mo><mn>1</mn></mrow></msub><mo>,</mo><mi>μ</mi><mo>+</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mfrac><mrow><mi>c</mi><mi>o</mi><mi>v</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mn>1</mn></mrow></msub><mo>,</mo><mi>Z</mi><msub><mi>α</mi><mrow><mn>1</mn><mo>,</mo><mn>2</mn></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><msubsup><mi>σ</mi><mi>α</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mfrac><mn>0</mn><mrow><msubsup><mi>σ</mi><mi>α</mi><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="center" style="text-align: center"><mo>=</mo></mtd><mtd columnalign="left" style="text-align: left"><mn>0</mn></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
  &amp;=&amp; cor( y_{1,1}, y_{1,2}) \\
      &amp;=&amp; cor( \mu + Z \alpha_{1,1} + e_{1,1}, \mu + Z \alpha_{1,2}  + e_{1,2}) \\
      &amp;=&amp; \frac{cov(Z \alpha_{1,1}, Z \alpha_{1,2})}{\sigma^2_\alpha + \sigma^2_\varepsilon} \\
      &amp;=&amp; \frac{0}{\sigma^2_\alpha + \sigma^2_\varepsilon} \\
      &amp;=&amp; 0
\end{eqnarray}</annotation></semantics></math> This interpretation in
terms of fraction of variation explained (FVE) naturally generalizes to
multiple variance components. Consider two sources of variation,
individual and cell type with variances
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{id}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>,</mo></mrow><annotation encoding="application/x-tex">\sigma^2_{cell},</annotation></semantics></math>
respectively. Applying a generalization of the the previous derivation,
two samples are correlated according to:</p>
<table>
<colgroup>
<col width="18%" />
<col width="14%" />
<col width="13%" />
<col width="26%" />
<col width="27%" />
</colgroup>
<thead>
<tr class="header">
<th>Individual</th>
<th>cell type</th>
<th>variance</th>
<th>Interpretation</th>
<th>Correlation value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>same</td>
<td>different</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mrow><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><annotation encoding="application/x-tex">\frac{\sigma^2_{id}}{\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon }</annotation></semantics></math></td>
<td>FVE by individual</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mi>C</mi><msub><mi>C</mi><mrow><mi>i</mi><mi>n</mi><mi>d</mi><mi>i</mi><mi>v</mi><mi>i</mi><mi>d</mi><mi>u</mi><mi>a</mi><mi>l</mi></mrow></msub></mrow><annotation encoding="application/x-tex">ICC_{individual}</annotation></semantics></math></td>
</tr>
<tr class="even">
<td>different</td>
<td>same</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mrow><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><annotation encoding="application/x-tex">\frac{\sigma^2_{cell}}{\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon }</annotation></semantics></math></td>
<td>FVE by cell type</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mi>C</mi><msub><mi>C</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow></msub></mrow><annotation encoding="application/x-tex">ICC_{cell}</annotation></semantics></math></td>
</tr>
<tr class="odd">
<td>same</td>
<td>same</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mrow><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup></mrow><mrow><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><annotation encoding="application/x-tex">\frac{\sigma^2_{id} + \sigma^2_{cell}}{\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon }</annotation></semantics></math></td>
<td>sum of FVE by individual &amp; cell type</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mi>C</mi><msub><mi>C</mi><mrow><mi>i</mi><mi>n</mi><mi>d</mi><mi>i</mi><mi>v</mi><mi>i</mi><mi>d</mi><mi>u</mi><mi>a</mi><mi>l</mi><mo>,</mo><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow></msub></mrow><annotation encoding="application/x-tex">ICC_{individual,cell}</annotation></semantics></math></td>
</tr>
<tr class="even">
<td>different</td>
<td>different</td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac><mn>0</mn><mrow><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow></mfrac><annotation encoding="application/x-tex">\frac{0}{\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon }</annotation></semantics></math></td>
<td>sample are independent</td>
<td></td>
</tr>
</tbody>
</table>
<p>Notice that the correlation between samples from the same individual
and same cell type corresponds to the sum of the fraction explained by
individual + fraction explained by cell type. This defines ICC for
individual and tissue, as well as the combined ICC and relates these
values to FVE.</p>
<p>In order to illustrate how this FVE and ICC relate to the correlation
between samples in multilevel datasets, consider a simple example of 5
samples from 2 individuals and 2 tissues:</p>
<table>
<thead>
<tr class="header">
<th>Sample</th>
<th>Individual</th>
<th>Cell type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>a</td>
<td>1</td>
<td>T-Cell</td>
</tr>
<tr class="even">
<td>b</td>
<td>1</td>
<td>T-Cell</td>
</tr>
<tr class="odd">
<td>c</td>
<td>1</td>
<td>monocyte</td>
</tr>
<tr class="even">
<td>d</td>
<td>2</td>
<td>T-Cell</td>
</tr>
<tr class="odd">
<td>e</td>
<td>2</td>
<td>monocyte</td>
</tr>
</tbody>
</table>
<p>Modeling the separate effects of individual and tissue gives the
following covariance structure between samples when a linear mixed model
is used:</p>
<div class="math">
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mi>c</mi><mi>o</mi><mi>v</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>y</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mtable><mtr><mtd columnalign="center" style="text-align: center"><mi>a</mi></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mi>b</mi></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mi>c</mi></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mi>d</mi></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mi>e</mi></mtd></mtr></mtable></mtd><mtd columnalign="center" style="text-align: center"><mrow><mo stretchy="true" form="prefix">(</mo><mtable><mtr><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"></mtd><mtd columnalign="center" style="text-align: center"></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><mn>0</mn></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"><mn>0</mn></mtd><mtd columnalign="center" style="text-align: center"><mn>0</mn></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup></mtd><mtd columnalign="center" style="text-align: center"><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mtd></mtr><mtr><mtd columnalign="center" style="text-align: center"></mtd></mtr></mtable><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{array}{ccc} &amp;
\\
cov(y)=\begin{array}{ccccc}
a \\
b \\
c\\
d\\
e\end{array}
&amp;
\left(
\begin{array}{ccccc}
\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon &amp; &amp; &amp;&amp;\\
\sigma^2_{id} + \sigma^2_{cell} &amp; \sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon  &amp; &amp; &amp;\\
\sigma^2_{id}  &amp; \sigma^2_{id} &amp;\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon  &amp; &amp;\\
\sigma^2_{cell} &amp; \sigma^2_{cell} &amp;0 &amp;\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon  &amp;\\
 0 &amp; 0 &amp;\sigma^2_{cell} &amp; \sigma^2_{id} &amp;\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon \\
\end{array}
\right)\end{array}
</annotation></semantics></math></p>
</div>
<p>The covariance matrix is symmetric so that blank entries take the
value on the opposite side of the diagonal. The covariance can be
converted to correlation by dividing by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>σ</mi><mrow><mi>i</mi><mi>d</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mrow><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi></mrow><mn>2</mn></msubsup><mo>+</mo><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup></mrow><annotation encoding="application/x-tex">\sigma^2_{id} + \sigma^2_{cell} + \sigma^2_\varepsilon</annotation></semantics></math>,
and this gives the results from above. This example generalizes to any
number of variance components <span class="citation">(Pinheiro and Bates
2000)</span>.</p>
</div>
<div id="variation-with-multiple-subsets-of-the-data"
class="section level3">
<h3>Variation with multiple subsets of the data</h3>
<p>The linear mixed model underlying <code>variancePartition</code>
allows the effect of one variable to depend on the value of another
variable. Statistically, this is called a varying coefficient model
<span class="citation">(Pinheiro and Bates 2000; Galecki and Burzykowski
2013)</span>. This model arises in <code>variancePartition</code>
analysis when the variation explained by individual depends on tissue or
cell type.</p>
<p>A given sample is only from one cell type, so this analysis asks a
question about a subset of the data. The the data is implicitly divided
into subsets base on cell type and variation explained by individual is
evaluated within each subset. The data is not actually divided onto
subset, but the statistical model essentially examples samples with each
cell type. This subsetting means that the variance fractions do not sum
to 1.</p>
<p>Consider a concrete example with variation from across individual and
cell types (T-cells and monocytes) with data from the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>i</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">i^{th}</annotation></semantics></math>
sample from the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>k</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup><annotation encoding="application/x-tex">k^{th}</annotation></semantics></math>
individual, sex of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
and cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>c</mi><annotation encoding="application/x-tex">c</annotation></semantics></math>.
Modeling the variation across individuals within cell type corresponds
to
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><msub><mi>y</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi><mo>,</mo><mi>s</mi><mo>,</mo><mi>c</mi></mrow></msub><mo>=</mo><mi>μ</mi><mo>+</mo><msup><mi>Z</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>s</mi><mi>e</mi><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><msub><mi>α</mi><mrow><mi>i</mi><mo>,</mo><mi>s</mi></mrow></msub><mo>+</mo><msup><mi>Z</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>T</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mo stretchy="false" form="prefix">|</mo><mi>i</mi><mi>d</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><msub><mi>α</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi><mo>,</mo><mi>c</mi></mrow></msub><mo>+</mo><msup><mi>Z</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mi>o</mi><mi>n</mi><mi>o</mi><mi>c</mi><mi>y</mi><mi>t</mi><mi>e</mi><mo stretchy="false" form="prefix">|</mo><mi>i</mi><mi>d</mi><mo stretchy="true" form="postfix">)</mo></mrow></msup><msub><mi>α</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi><mo>,</mo><mi>c</mi></mrow></msub><mo>+</mo><msub><mi>e</mi><mrow><mi>i</mi><mo>,</mo><mi>k</mi><mo>,</mo><mi>s</mi><mo>,</mo><mi>c</mi></mrow></msub></mtd></mtr></mtable><annotation encoding="application/x-tex">\begin{eqnarray}
  y_{i,k,s,c} = \mu + Z^{(sex)} \alpha_{i,s} + Z^{(Tcell| id)} \alpha_{i,k,c} + Z^{(monocyte| id)} \alpha_{i,k,c} + e_{i,k,s,c}
\end{eqnarray}</annotation></semantics></math></p>
<p>with corresponding variance components:</p>
<table>
<colgroup>
<col width="56%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th>Variance component</th>
<th>Interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mrow><mi>s</mi><mi>e</mi><mi>x</mi></mrow><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{sex}</annotation></semantics></math></td>
<td>variance across sex (which is the same for all cell types)</td>
</tr>
<tr class="even">
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>T</mi><mi>c</mi><mi>e</mi><mi>l</mi><mi>l</mi><mo stretchy="false" form="prefix">|</mo><mi>i</mi><mi>d</mi><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{(Tcell| id)}</annotation></semantics></math></td>
<td>variance across individuals within T-cells</td>
</tr>
<tr class="odd">
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mi>o</mi><mi>n</mi><mi>o</mi><mi>c</mi><mi>y</mi><mi>t</mi><mi>e</mi><mo stretchy="false" form="prefix">|</mo><mi>i</mi><mi>d</mi><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_{(monocyte| id)}</annotation></semantics></math></td>
<td>variance across individuals within monocytes</td>
</tr>
<tr class="even">
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>ε</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma^2_\varepsilon</annotation></semantics></math></td>
<td>residual variance</td>
</tr>
</tbody>
</table>
<p>Since the dataset is now divided into multiple subsets, direct
interpretation of the fraction of variation explained (FVE) as
intra-class correlation does not apply. Instead, we compute a
“pseudo-FVE” by approximating the total variance attributable to cell
type by using a weighted average of the within cell type variances
weighted by the sample size within each cell type. Thus the values of
pseudo-FVE do not have the simple interpretation as in the standard
application of <code>variancePartition</code>, but allows ranking of
variables based on genome-wide contribution to variance and analysis of
gene-level results.</p>
</div>
</div>
<div id="variance-partitioning-and-differential-expression"
class="section level2">
<h2>Variance partitioning and differential expression</h2>
<p>Differential expression (DE) is widely used to identify gene which
show difference is expression between two subsets of the data (i.e. case
versus controls). For a single gene, DE analysis measures the difference
in mean expression between the two subsets. (Since expression is usually
analyzed on a log scale, DE results are usually shown in terms of log
fold changes between the two subsets ). In the Figure, consider two
simulated examples of a gene whose expression differs between males and
females. The mean expression in males is 0 and the mean expression in
females is 2 in both cases. Therefore, the fold change is 2 in both
cases.</p>
<p>However, the fraction of expression variation explained by sex is
very different in these two examples. In example A, there is very little
variation <em>within</em> each sex, so that variation <em>between</em>
sexes is very high at 91.1%. Conversely, example B shows high variation
<em>within</em> sexes, so that variation <em>between</em> sexes is only
17.8%.</p>
<p>The fact that the fold change or the fraction of variation is
significantly different from 0 indicates differential expression between
the two sexes. Yet these two statistics have different interpretations.
The fold change from DE analysis tests a difference in means between two
sexes. The fraction of variation explained compares the variation
explained by sex to the total variation.</p>
<p>Thus the fraction of variation explained reported by
<code>variancePartition</code> reflects as different aspect of the data
not captured by DE analysis.</p>
<p><img src="/Users/gabrielhoffman/workspace/repos/variancePartition/docs/articles/variancePartition_files/figure-html/DE-1.png" width="768" style="display: block; margin: auto;" /></p>
<div id="modelling-error-in-gene-expression-measurements"
class="section level3">
<h3>Modelling error in gene expression measurements</h3>
<p>Uncertainty in the measurement of gene expression can be modeled with
precision weights and tests of differentially expression using
<code>limma::voom()</code> model this uncertainty directly with a
heteroskedastic linear regression <span class="citation">(Law et al.
2014)</span>. <code>variancePartition</code> can use these precision
weights in a heteroskedastic linear mixed model implemented in
<code>lme4</code> <span class="citation">(Bates et al. 2015)</span>.
These precision weights are used seamlessly by calling
<code>fitVarPartModel()</code> or <code>fitExtractVarPartModel()</code>
on the output of <code>limma::voom()</code>. Otherwise the user can
specify the weights with the <code>weightsMatrix</code> parameter.</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<details>
<pre><code>## R version 4.3.0 (2023-04-21)
## Platform: x86_64-apple-darwin22.4.0 (64-bit)
## Running under: macOS 14.2.1
## 
## Matrix products: default
## BLAS:   /Users/gabrielhoffman/prog/R-4.3.0/lib/libRblas.dylib 
## LAPACK: /usr/local/Cellar/r/4.3.0_1/lib/R/lib/libRlapack.dylib;  LAPACK version 3.11.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/New_York
## tzcode source: internal
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] dendextend_1.17.1           ballgown_2.32.0             DESeq2_1.40.1              
##  [4] SummarizedExperiment_1.30.1 Biobase_2.60.0              MatrixGenerics_1.12.0      
##  [7] matrixStats_1.0.0           GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        
## [10] IRanges_2.34.1              S4Vectors_0.38.1            BiocGenerics_0.46.0        
## [13] edgeR_3.42.4                lme4_1.1-34                 Matrix_1.5-4.1             
## [16] variancePartition_1.33.11   BiocParallel_1.34.2         limma_3.56.2               
## [19] ggplot2_3.4.4              
## 
## loaded via a namespace (and not attached):
##   [1] RColorBrewer_1.1-3       jsonlite_1.8.5           magrittr_2.0.3          
##   [4] farver_2.1.1             nloptr_2.0.3             rmarkdown_2.22          
##   [7] fs_1.6.2                 BiocIO_1.10.0            zlibbioc_1.46.0         
##  [10] ragg_1.2.5               vctrs_0.6.3              memoise_2.0.1           
##  [13] minqa_1.2.5              Rsamtools_2.16.0         RCurl_1.98-1.12         
##  [16] htmltools_0.5.5          S4Arrays_1.2.0           broom_1.0.5             
##  [19] sass_0.4.6               KernSmooth_2.23-21       bslib_0.4.2             
##  [22] desc_1.4.2               pbkrtest_0.5.2           plyr_1.8.8              
##  [25] cachem_1.0.8             GenomicAlignments_1.36.0 lifecycle_1.0.3         
##  [28] iterators_1.0.14         pkgconfig_2.0.3          R6_2.5.1                
##  [31] fastmap_1.1.1            GenomeInfoDbData_1.2.10  rbibutils_2.2.13        
##  [34] digest_0.6.33            numDeriv_2016.8-1.1      colorspace_2.1-0        
##  [37] AnnotationDbi_1.62.1     rprojroot_2.0.3          textshaping_0.3.6       
##  [40] RSQLite_2.3.1            labeling_0.4.2           fansi_1.0.4             
##  [43] mgcv_1.8-42              httr_1.4.6               abind_1.4-5             
##  [46] compiler_4.3.0           bit64_4.0.5              aod_1.3.2               
##  [49] withr_2.5.0              backports_1.4.1          viridis_0.6.3           
##  [52] DBI_1.1.3                highr_0.10               gplots_3.1.3            
##  [55] MASS_7.3-60              DelayedArray_0.26.3      rjson_0.2.21            
##  [58] corpcor_1.6.10           gtools_3.9.4             caTools_1.18.2          
##  [61] tools_4.3.0              remaCor_0.0.17           glue_1.6.2              
##  [64] restfulr_0.0.15          nlme_3.1-162             grid_4.3.0              
##  [67] reshape2_1.4.4           generics_0.1.3           sva_3.48.0              
##  [70] gtable_0.3.3             tidyr_1.3.0              utf8_1.2.3              
##  [73] XVector_0.40.0           pillar_1.9.0             stringr_1.5.0           
##  [76] genefilter_1.82.1        splines_4.3.0            dplyr_1.1.2             
##  [79] lattice_0.21-8           survival_3.5-5           rtracklayer_1.60.0      
##  [82] bit_4.0.5                annotate_1.78.0          tidyselect_1.2.0        
##  [85] locfit_1.5-9.7           Biostrings_2.68.1        knitr_1.43              
##  [88] gridExtra_2.3            RhpcBLASctl_0.23-42      xfun_0.39               
##  [91] stringi_1.7.12           yaml_2.3.7               boot_1.3-28.1           
##  [94] evaluate_0.21            codetools_0.2-19         tibble_3.2.1            
##  [97] cli_3.6.1                xtable_1.8-4             systemfonts_1.0.4       
## [100] Rdpack_2.4               munsell_0.5.0            jquerylib_0.1.4         
## [103] Rcpp_1.0.11              EnvStats_2.7.0           png_0.1-8               
## [106] XML_3.99-0.14            parallel_4.3.0           pkgdown_2.0.7           
## [109] blob_1.2.4               bitops_1.0-7             viridisLite_0.4.2       
## [112] mvtnorm_1.2-2            lmerTest_3.1-3           scales_1.2.1            
## [115] purrr_1.0.2              crayon_1.5.2             fANCOVA_0.6-1           
## [118] rlang_1.1.1              cowplot_1.1.1            KEGGREST_1.40.0</code></pre>
</details>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent"
entry-spacing="0">
<div id="ref-Anders2015" class="csl-entry">
Anders, Simon, Paul Theodor Pyl, and Wolfgang Huber. 2015. <span>“<span
class="nocase">HTSeq-A Python framework to work with high-throughput
sequencing data</span>.”</span> <em>Bioinformatics</em> 31 (2): 166–69.
<a
href="https://doi.org/10.1093/bioinformatics/btu638">https://doi.org/10.1093/bioinformatics/btu638</a>.
</div>
<div id="ref-Bates2015" class="csl-entry">
Bates, D., M. Machler, B. Bolker, and S. Walker. 2015. <span>“<span
class="nocase">Fitting Linear Mixed-Effects Models Using
lme4</span>.”</span> <em>Journal of Statistical Software</em> 67 (1). <a
href="https://doi.org/10.18637/jss.v067.i01">https://doi.org/10.18637/jss.v067.i01</a>.
</div>
<div id="ref-Bray2015" class="csl-entry">
Bray, Nicolas L, Harold Pimentel, Páll Melsted, and Lior Pachter. 2016.
<span>“<span class="nocase">Near-optimal probabilistic RNA-seq
quantification</span>.”</span> <em>Nature Biotechnology</em> 34: 525–27.
<a
href="https://doi.org/10.1038/nbt.3519">https://doi.org/10.1038/nbt.3519</a>.
</div>
<div id="ref-oligoBioc" class="csl-entry">
Carvalho, Benilton S, and Rafael A Irizarry. 2010. <span>“A Framework
for Oligonucleotide Microarray Preprocessing.”</span>
<em>Bioinformatics</em> 26 (19): 2363–67. <a
href="https://doi.org/10.1093/bioinformatics/btq431">https://doi.org/10.1093/bioinformatics/btq431</a>.
</div>
<div id="ref-Frazee2015" class="csl-entry">
Frazee, Alyssa C, Geo Pertea, Andrew E Jaffe, Ben Langmead, Steven L
Salzberg, and Jeffrey T Leek. 2015. <span>“<span class="nocase">Ballgown
bridges the gap between transcriptome assembly and expression
analysis</span>.”</span> <em>Nature Biotechnology</em> 33 (3): 243–46.
<a
href="https://doi.org/10.1038/nbt.3172">https://doi.org/10.1038/nbt.3172</a>.
</div>
<div id="ref-Galecki2010" class="csl-entry">
Galecki, A., and T. Burzykowski. 2013. <em><span class="nocase">Linear
Mixed Effects Modeling using R.</span></em> Springer.
</div>
<div id="ref-Law2014" class="csl-entry">
Law, C. W., Y. Chen, W. Shi, and G. K. Smyth. 2014. <span>“<span
class="nocase">Voom: precision weights unlock linear model analysis
tools for RNA-seq read counts.</span>”</span> <em>Genome Biology</em> 15
(2): R29. <a
href="https://doi.org/10.1186/gb-2014-15-2-r29">https://doi.org/10.1186/gb-2014-15-2-r29</a>.
</div>
<div id="ref-Li2011f" class="csl-entry">
Li, Bo, and Colin N Dewey. 2011. <span>“<span class="nocase">RSEM:
accurate transcript quantification from RNA-Seq data with or without a
reference genome</span>.”</span> <em>BMC Bioinformatics</em> 12 (1):
323. <a
href="https://doi.org/10.1186/1471-2105-12-323">https://doi.org/10.1186/1471-2105-12-323</a>.
</div>
<div id="ref-Liao2014" class="csl-entry">
Liao, Yang, Gordon K. Smyth, and Wei Shi. 2014. <span>“<span
class="nocase">FeatureCounts: An efficient general purpose program for
assigning sequence reads to genomic features</span>.”</span>
<em>Bioinformatics</em> 30 (7): 923–30. <a
href="https://doi.org/10.1093/bioinformatics/btt656">https://doi.org/10.1093/bioinformatics/btt656</a>.
</div>
<div id="ref-Love2014" class="csl-entry">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“<span
class="nocase">Moderated estimation of fold change and dispersion for
RNA-seq data with DESeq2</span>.”</span> <em>Genome Biology</em> 15
(12): 550. <a
href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Patro2015" class="csl-entry">
Patro, Rob, Geet Duggal, and Carl Kingsford. 2015. <span>“<span
class="nocase">Accurate, fast, and model-aware transcript expression
quantification with Salmon</span>.”</span> <em>Bioarxiv</em>. <a
href="https://doi.org/10.1101/021592">https://doi.org/10.1101/021592</a>.
</div>
<div id="ref-Patro2014" class="csl-entry">
Patro, Rob, Stephen M Mount, and Carl Kingsford. 2014. <span>“<span
class="nocase">Sailfish enables alignment-free isoform quantification
from RNA-seq reads using lightweight algorithms</span>.”</span>
<em>Nature Biotechnology</em> 32 (5): 462–64. <a
href="https://doi.org/10.1038/nbt.2862">https://doi.org/10.1038/nbt.2862</a>.
</div>
<div id="ref-Pinheiro2000" class="csl-entry">
Pinheiro, J. C., and D. M. Bates. 2000. <em><a
href="https://www.ncbi.nlm.nih.gov/pubmed/40509311"><span
class="nocase">Mixed-Effects Models in S and S-Plus</span></a></em>. New
York: Springer.
</div>
<div id="ref-Robinson2010c" class="csl-entry">
Robinson, M. D., and A. Oshlack. 2010. <span>“<span class="nocase">A
scaling normalization method for differential expression analysis of
RNA-seq data.</span>”</span> <em>Genome Biology</em> 11 (3): R25. <a
href="https://doi.org/10.1186/gb-2010-11-3-r25">https://doi.org/10.1186/gb-2010-11-3-r25</a>.
</div>
<div id="ref-Soneson2015" class="csl-entry">
Soneson, Charlotte, Michael I. Love, and Mark D. Robinson. 2015.
<span>“<span class="nocase">Differential analyses for RNA-seq:
transcript-level estimates improve gene-level inferences</span>.”</span>
<em>F1000Research</em> 4 (0): 1521. <a
href="https://doi.org/10.12688/f1000research.7563.2">https://doi.org/10.12688/f1000research.7563.2</a>.
</div>
<div id="ref-Trapnell2010" class="csl-entry">
Trapnell, Cole, Brian a Williams, Geo Pertea, Ali Mortazavi, Gordon
Kwan, Marijke J van Baren, Steven L Salzberg, Barbara J Wold, and Lior
Pachter. 2010. <span>“<span class="nocase">Transcript assembly and
quantification by RNA-Seq reveals unannotated transcripts and isoform
switching during cell differentiation</span>.”</span> <em>Nature
Biotechnology</em> 28 (5): 511–15. <a
href="https://doi.org/10.1038/nbt.1621">https://doi.org/10.1038/nbt.1621</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>
