<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Transform RNA-Seq Data Ready for Linear Mixed Modelling with dream() — voomWithDreamWeights • variancePartition</title><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Transform RNA-Seq Data Ready for Linear Mixed Modelling with dream() — voomWithDreamWeights"><meta name="description" content="Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with dream().   This method is the same as limma::voom(), except that it allows random effects in the formula"><meta property="og:description" content="Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with dream().   This method is the same as limma::voom(), except that it allows random effects in the formula"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.39.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/variancePartition.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dream.html">dream analysis</a></li>
    <li><a class="dropdown-item" href="../articles/rnd_effects.html">Theory and practice of random effects</a></li>
    <li><a class="dropdown-item" href="../articles/mvtests.html">Multivariate tests</a></li>
    <li><a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Transform RNA-Seq Data Ready for Linear Mixed Modelling with <code>dream()</code></h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/R/voomWithDreamWeights.R" class="external-link"><code>R/voomWithDreamWeights.R</code></a></small>
      <div class="d-none name"><code>voomWithDreamWeights.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with <code><a href="dream-method.html">dream()</a></code>.   This method is the same as <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code>, except that it allows random effects in the formula</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">voomWithDreamWeights</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  <span class="va">formula</span>,</span>
<span>  <span class="va">data</span>,</span>
<span>  lib.size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize.method <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  span <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior.count <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  prior.count.for.weights <span class="op">=</span> <span class="va">prior.count</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  save.plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rescaleWeightsAfter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scaledByLib <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  priorWeightsAsCounts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>a numeric <code>matrix</code> containing raw counts, or an <code>ExpressionSet</code> containing raw counts, or a <code>DGEList</code> object. Counts must be non-negative and NAs are not permitted.</p></dd>


<dt id="arg-formula">formula<a class="anchor" aria-label="anchor" href="#arg-formula"></a></dt>
<dd><p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used as a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work, and <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit()</a></code> followed by contrasts.fit() are run.</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p><code>data.frame</code> with columns corresponding to formula</p></dd>


<dt id="arg-lib-size">lib.size<a class="anchor" aria-label="anchor" href="#arg-lib-size"></a></dt>
<dd><p>numeric vector containing total library sizes for each sample.  Defaults to the normalized (effective) library sizes in <code>counts</code> if <code>counts</code> is a <code>DGEList</code> or to the columnwise count totals if <code>counts</code> is a matrix.</p></dd>


<dt id="arg-normalize-method">normalize.method<a class="anchor" aria-label="anchor" href="#arg-normalize-method"></a></dt>
<dd><p>the microarray-style normalization method to be applied to the logCPM values (if any).  Choices are as for the <code>method</code> argument of <code>normalizeBetweenArrays</code> when the data is single-channel.  Any normalization factors found in <code>counts</code> will still be used even if <code>normalize.method="none"</code>.</p></dd>


<dt id="arg-span">span<a class="anchor" aria-label="anchor" href="#arg-span"></a></dt>
<dd><p>width of the lowess smoothing window as a proportion. Setting <code>span="auto"</code> uses <code><a href="https://rdrr.io/pkg/fANCOVA/man/loess.as.html" class="external-link">fANCOVA::loess.as()</a></code> to estimate the tuning parameter from the data</p></dd>


<dt id="arg-weights">weights<a class="anchor" aria-label="anchor" href="#arg-weights"></a></dt>
<dd><p>Can be a numeric matrix of individual weights of same dimensions as the <code>counts</code>, or a numeric vector of sample weights with length equal to <code>ncol(counts)</code></p></dd>


<dt id="arg-prior-count">prior.count<a class="anchor" aria-label="anchor" href="#arg-prior-count"></a></dt>
<dd><p>average count to be added to each observation to avoid taking log of zero. The count applied to each sample is normalized by library size so given equal log CPM for a gene with zero counts across multiple samples</p></dd>


<dt id="arg-prior-count-for-weights">prior.count.for.weights<a class="anchor" aria-label="anchor" href="#arg-prior-count-for-weights"></a></dt>
<dd><p>count added to regularize weights</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>logical, should a plot of the mean-variance trend be displayed?</p></dd>


<dt id="arg-save-plot">save.plot<a class="anchor" aria-label="anchor" href="#arg-save-plot"></a></dt>
<dd><p>logical, should the coordinates and line of the plot be saved in the output?</p></dd>


<dt id="arg-rescaleweightsafter">rescaleWeightsAfter<a class="anchor" aria-label="anchor" href="#arg-rescaleweightsafter"></a></dt>
<dd><p>default = FALSE, should the output weights be scaled by the input weights</p></dd>


<dt id="arg-scaledbylib">scaledByLib<a class="anchor" aria-label="anchor" href="#arg-scaledbylib"></a></dt>
<dd><p>if <code>TRUE</code>, scale pseudocount by <code>lib.size</code>.  Else to standard constant pseudocount addition</p></dd>


<dt id="arg-priorweightsascounts">priorWeightsAsCounts<a class="anchor" aria-label="anchor" href="#arg-priorweightsascounts"></a></dt>
<dd><p>if <code>weights</code> is <code>NULL</code>, set weights to be equal to counts, following delta method for log2 CPM</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>parameters for parallel evaluation</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other arguments are passed to <code>lmer</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An <code>EList</code> object just like the result of <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code></p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Adapted from <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom()</a></code> in <code>limma</code> v3.40.2</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># library(variancePartition)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">varPartDEdata</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># normalize RNA-seq counts</span></span></span>
<span class="r-in"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">countMatrix</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># specify formula with random effect for Individual</span></span></span>
<span class="r-in"><span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Disease</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Individual</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># compute observation weights</span></span></span>
<span class="r-in"><span><span class="va">vobj</span> <span class="op">&lt;-</span> <span class="fu">voomWithDreamWeights</span><span class="op">(</span><span class="va">dge</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="op">]</span>, <span class="va">form</span>, <span class="va">metadata</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># fit dream model</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobj</span>, <span class="va">form</span>, <span class="va">metadata</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="eBayes.html">eBayes</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="toptable-method.html">topTable</a></span><span class="op">(</span><span class="va">res</span>, coef <span class="op">=</span> <span class="st">"Disease1"</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                    logFC  AveExpr        t      P.Value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000456159.1 gene=MET     1.0182945 2.458926 6.241638 7.270470e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000418210.2 gene=TMEM64  1.0375652 4.715367 6.424903 3.175466e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000555834.1 gene=RPS6KL1 0.9355651 5.272063 5.653604 3.749850e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                   adj.P.Val        B    z.std</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000456159.1 gene=MET     1.454094e-05 5.834316 4.953996</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000418210.2 gene=TMEM64  2.499900e-05 5.811378 4.659131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENST00000555834.1 gene=RPS6KL1 2.499900e-05 4.223847 4.624786</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

